{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bd2c3374",
   "metadata": {},
   "source": [
    "## Sourcing data from NCBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0f53d0d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ./datasets summary taxonomy taxon \"Pectobacteriaceae\"\n",
    "P_taxon_id = 1903410"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45c6ed55",
   "metadata": {},
   "source": [
    "## Importing data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc8a709a",
   "metadata": {},
   "source": [
    "### Phage accession lookup\n",
    "Phage name, ID, NCBI accession "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2c20ab7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phage_name</th>\n",
       "      <th>id</th>\n",
       "      <th>accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>244</td>\n",
       "      <td>DQ398041</td>\n",
       "      <td>NC_008194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>32HC</td>\n",
       "      <td>KJ028219</td>\n",
       "      <td>NC_023602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>39HC</td>\n",
       "      <td>KJ433973</td>\n",
       "      <td>NC_023603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40AC</td>\n",
       "      <td>KJ192196</td>\n",
       "      <td>NC_023607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40BC</td>\n",
       "      <td>KJ433975</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5073</th>\n",
       "      <td>Zorro</td>\n",
       "      <td>MK279896</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5074</th>\n",
       "      <td>Zucker</td>\n",
       "      <td>ON645346</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5075</th>\n",
       "      <td>Zuko</td>\n",
       "      <td>MN204493</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5076</th>\n",
       "      <td>Zulu</td>\n",
       "      <td>MH779517</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5077</th>\n",
       "      <td>ZygoTaiga</td>\n",
       "      <td>KM881426</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5078 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     phage_name        id  accession\n",
       "0           244  DQ398041  NC_008194\n",
       "1          32HC  KJ028219  NC_023602\n",
       "2          39HC  KJ433973  NC_023603\n",
       "3          40AC  KJ192196  NC_023607\n",
       "4          40BC  KJ433975        NaN\n",
       "...         ...       ...        ...\n",
       "5073      Zorro  MK279896        NaN\n",
       "5074     Zucker  ON645346        NaN\n",
       "5075       Zuko  MN204493        NaN\n",
       "5076       Zulu  MH779517        NaN\n",
       "5077  ZygoTaiga  KM881426        NaN\n",
       "\n",
       "[5078 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import sys\n",
    "os_system = sys.platform #recognize operationg system.\n",
    "if os_system == \"win32\":\n",
    "    datasets = \"./datasets.exe\"\n",
    "    dataformat = \"./dataformat.exe\"\n",
    "else:\n",
    "    datasets = \"./datasets_macOS\"\n",
    "    dataformat = \"./dataformat_macOS\"\n",
    "\n",
    "PhagesDB_all = pd.read_csv(\"PhagesDB_accession_numbers.txt\", sep=\"\\t\", names=[\"phage_name\", \"id\", \"accession\"], header=0)\n",
    "PhagesDB_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb5a0d71",
   "metadata": {},
   "source": [
    "Only phages with accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ac4cf64f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phage_name</th>\n",
       "      <th>id</th>\n",
       "      <th>accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>244</td>\n",
       "      <td>DQ398041</td>\n",
       "      <td>NC_008194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>32HC</td>\n",
       "      <td>KJ028219</td>\n",
       "      <td>NC_023602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>39HC</td>\n",
       "      <td>KJ433973</td>\n",
       "      <td>NC_023603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40AC</td>\n",
       "      <td>KJ192196</td>\n",
       "      <td>NC_023607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Acadian</td>\n",
       "      <td>JN699007</td>\n",
       "      <td>NC_023701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4917</th>\n",
       "      <td>WIVsmall</td>\n",
       "      <td>KC736071</td>\n",
       "      <td>NC_021334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4945</th>\n",
       "      <td>Xeno</td>\n",
       "      <td>KU935728</td>\n",
       "      <td>NC_031243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5019</th>\n",
       "      <td>Zaka</td>\n",
       "      <td>KF560334</td>\n",
       "      <td>NC_022985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5041</th>\n",
       "      <td>Zemlya</td>\n",
       "      <td>KC700558</td>\n",
       "      <td>NC_021339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5066</th>\n",
       "      <td>ZoeJ</td>\n",
       "      <td>KJ510412</td>\n",
       "      <td>NC_024147</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>287 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     phage_name        id  accession\n",
       "0           244  DQ398041  NC_008194\n",
       "1          32HC  KJ028219  NC_023602\n",
       "2          39HC  KJ433973  NC_023603\n",
       "3          40AC  KJ192196  NC_023607\n",
       "24      Acadian  JN699007  NC_023701\n",
       "...         ...       ...        ...\n",
       "4917   WIVsmall  KC736071  NC_021334\n",
       "4945       Xeno  KU935728  NC_031243\n",
       "5019       Zaka  KF560334  NC_022985\n",
       "5041     Zemlya  KC700558  NC_021339\n",
       "5066       ZoeJ  KJ510412  NC_024147\n",
       "\n",
       "[287 rows x 3 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PDB_acc = PhagesDB_all[PhagesDB_all['accession'].notna()]\n",
    "PDB_acc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "892b112d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phage_name</th>\n",
       "      <th>id</th>\n",
       "      <th>accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>244</td>\n",
       "      <td>DQ398041</td>\n",
       "      <td>NC_008194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>32HC</td>\n",
       "      <td>KJ028219</td>\n",
       "      <td>NC_023602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>39HC</td>\n",
       "      <td>KJ433973</td>\n",
       "      <td>NC_023603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40AC</td>\n",
       "      <td>KJ192196</td>\n",
       "      <td>NC_023607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Acadian</td>\n",
       "      <td>JN699007</td>\n",
       "      <td>NC_023701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Adawi</td>\n",
       "      <td>KF279411</td>\n",
       "      <td>NC_022328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Adjutor</td>\n",
       "      <td>EU676000</td>\n",
       "      <td>NC_010763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Adzzy</td>\n",
       "      <td>KF416344</td>\n",
       "      <td>NC_022058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Aeneas</td>\n",
       "      <td>JQ809703</td>\n",
       "      <td>NC_023723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Akoma</td>\n",
       "      <td>JN699006</td>\n",
       "      <td>NC_023742</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   phage_name        id  accession\n",
       "0         244  DQ398041  NC_008194\n",
       "1        32HC  KJ028219  NC_023602\n",
       "2        39HC  KJ433973  NC_023603\n",
       "3        40AC  KJ192196  NC_023607\n",
       "24    Acadian  JN699007  NC_023701\n",
       "36      Adawi  KF279411  NC_022328\n",
       "42    Adjutor  EU676000  NC_010763\n",
       "51      Adzzy  KF416344  NC_022058\n",
       "54     Aeneas  JQ809703  NC_023723\n",
       "77      Akoma  JN699006  NC_023742"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(PDB_acc.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b99a3b98",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fetching data from NCBI: 100%|██████████| 287/287 [02:48<00:00,  1.70acc/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>bioprojects</th>\n",
       "      <th>completeness</th>\n",
       "      <th>host</th>\n",
       "      <th>is_annotated</th>\n",
       "      <th>is_lab_host</th>\n",
       "      <th>lab_host</th>\n",
       "      <th>length</th>\n",
       "      <th>nucleotide</th>\n",
       "      <th>protein_count</th>\n",
       "      <th>release_date</th>\n",
       "      <th>source_database</th>\n",
       "      <th>submitter</th>\n",
       "      <th>update_date</th>\n",
       "      <th>virus</th>\n",
       "      <th>isolate</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_008194.1</td>\n",
       "      <td>[PRJNA485481]</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>{'lineage': [{'name': 'cellular organisms', 't...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Mycobacterium smegmatis</td>\n",
       "      <td>74483</td>\n",
       "      <td>{'sequence_hash': '2974DF27'}</td>\n",
       "      <td>142</td>\n",
       "      <td>2006-06-20T00:00:00Z</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>{'affiliation': 'National Center for Biotechno...</td>\n",
       "      <td>2023-01-11T00:00:00Z</td>\n",
       "      <td>{'lineage': [{'name': 'Viruses', 'tax_id': 102...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_023602.1</td>\n",
       "      <td>[PRJNA485481]</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>{'lineage': [{'name': 'cellular organisms', 't...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50781</td>\n",
       "      <td>{'sequence_hash': 'CFEA6EA'}</td>\n",
       "      <td>86</td>\n",
       "      <td>2014-03-04T00:00:00Z</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>{'affiliation': 'National Center for Biotechno...</td>\n",
       "      <td>2023-01-08T00:00:00Z</td>\n",
       "      <td>{'lineage': [{'name': 'Viruses', 'tax_id': 102...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_023603.1</td>\n",
       "      <td>[PRJNA485481]</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>{'lineage': [{'name': 'cellular organisms', 't...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71565</td>\n",
       "      <td>{'sequence_hash': '52C4C232'}</td>\n",
       "      <td>100</td>\n",
       "      <td>2014-03-04T00:00:00Z</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>{'affiliation': 'National Center for Biotechno...</td>\n",
       "      <td>2023-01-08T00:00:00Z</td>\n",
       "      <td>{'lineage': [{'name': 'Viruses', 'tax_id': 102...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_023607.1</td>\n",
       "      <td>[PRJNA485481]</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>{'lineage': [{'name': 'cellular organisms', 't...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>53396</td>\n",
       "      <td>{'sequence_hash': '4EC4E167'}</td>\n",
       "      <td>90</td>\n",
       "      <td>2014-03-04T00:00:00Z</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>{'affiliation': 'National Center for Biotechno...</td>\n",
       "      <td>2023-01-08T00:00:00Z</td>\n",
       "      <td>{'lineage': [{'name': 'Viruses', 'tax_id': 102...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_023701.1</td>\n",
       "      <td>[PRJNA485481]</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>{'lineage': [{'name': 'cellular organisms', 't...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Mycobacterium smegmatis mc2 155</td>\n",
       "      <td>69864</td>\n",
       "      <td>{'sequence_hash': '4E62B368'}</td>\n",
       "      <td>97</td>\n",
       "      <td>2014-03-11T00:00:00Z</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>{'affiliation': 'National Center for Biotechno...</td>\n",
       "      <td>2023-01-11T00:00:00Z</td>\n",
       "      <td>{'lineage': [{'name': 'Viruses', 'tax_id': 102...</td>\n",
       "      <td>{'collection_date': '2009-07-01'}</td>\n",
       "      <td>{'geographic_location': 'USA: Lafayette, LA', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>NC_021334.1</td>\n",
       "      <td>[PRJNA485481]</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>{'lineage': [{'name': 'cellular organisms', 't...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Mycobacterium smegmatis mc2 155</td>\n",
       "      <td>53359</td>\n",
       "      <td>{'sequence_hash': 'A8A0EF4E'}</td>\n",
       "      <td>83</td>\n",
       "      <td>2013-06-03T00:00:00Z</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>{'affiliation': 'National Center for Biotechno...</td>\n",
       "      <td>2023-01-08T00:00:00Z</td>\n",
       "      <td>{'lineage': [{'name': 'Viruses', 'tax_id': 102...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'geographic_location': 'China', 'geographic_r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>NC_031243.1</td>\n",
       "      <td>[PRJNA485481]</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>{'lineage': [{'name': 'cellular organisms', 't...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Mycobacterium smegmatis mc2 155</td>\n",
       "      <td>42395</td>\n",
       "      <td>{'sequence_hash': '5E5F2E14'}</td>\n",
       "      <td>69</td>\n",
       "      <td>2016-09-27T00:00:00Z</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>{'affiliation': 'National Center for Biotechno...</td>\n",
       "      <td>2023-01-09T00:00:00Z</td>\n",
       "      <td>{'lineage': [{'name': 'Viruses', 'tax_id': 102...</td>\n",
       "      <td>{'collection_date': '2014-10-01'}</td>\n",
       "      <td>{'geographic_location': 'USA: New Haven, CT', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>NC_022985.1</td>\n",
       "      <td>[PRJNA485481]</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>{'lineage': [{'name': 'cellular organisms', 't...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Mycobacterium smegmatis mc2 155</td>\n",
       "      <td>52122</td>\n",
       "      <td>{'sequence_hash': '910ADD35'}</td>\n",
       "      <td>101</td>\n",
       "      <td>2013-12-04T00:00:00Z</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>{'affiliation': 'National Center for Biotechno...</td>\n",
       "      <td>2023-01-08T00:00:00Z</td>\n",
       "      <td>{'lineage': [{'name': 'Viruses', 'tax_id': 102...</td>\n",
       "      <td>{'collection_date': '2011-09-29'}</td>\n",
       "      <td>{'geographic_location': 'USA: Los Angeles, CA'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>NC_021339.1</td>\n",
       "      <td>[PRJNA485481]</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>{'lineage': [{'name': 'cellular organisms', 't...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Streptomyces lividans</td>\n",
       "      <td>51077</td>\n",
       "      <td>{'sequence_hash': '7D95B565'}</td>\n",
       "      <td>76</td>\n",
       "      <td>2013-06-03T00:00:00Z</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>{'affiliation': 'National Center for Biotechno...</td>\n",
       "      <td>2023-01-08T00:00:00Z</td>\n",
       "      <td>{'lineage': [{'name': 'Viruses', 'tax_id': 102...</td>\n",
       "      <td>{'collection_date': '2008-10-30'}</td>\n",
       "      <td>{'geographic_location': 'USA: Houston, Texas',...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>NC_024147.1</td>\n",
       "      <td>[PRJNA485481]</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>{'lineage': [{'name': 'cellular organisms', 't...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Mycobacterium smegmatis mc2 155</td>\n",
       "      <td>57315</td>\n",
       "      <td>{'sequence_hash': '7CA1F84F'}</td>\n",
       "      <td>92</td>\n",
       "      <td>2014-05-27T00:00:00Z</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>{'affiliation': 'National Center for Biotechno...</td>\n",
       "      <td>2023-01-08T00:00:00Z</td>\n",
       "      <td>{'lineage': [{'name': 'Viruses', 'tax_id': 102...</td>\n",
       "      <td>{'collection_date': '2012-10-01'}</td>\n",
       "      <td>{'geographic_location': 'USA', 'geographic_reg...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession    bioprojects completeness  \\\n",
       "0    NC_008194.1  [PRJNA485481]     COMPLETE   \n",
       "1    NC_023602.1  [PRJNA485481]     COMPLETE   \n",
       "2    NC_023603.1  [PRJNA485481]     COMPLETE   \n",
       "3    NC_023607.1  [PRJNA485481]     COMPLETE   \n",
       "4    NC_023701.1  [PRJNA485481]     COMPLETE   \n",
       "..           ...            ...          ...   \n",
       "258  NC_021334.1  [PRJNA485481]     COMPLETE   \n",
       "259  NC_031243.1  [PRJNA485481]     COMPLETE   \n",
       "260  NC_022985.1  [PRJNA485481]     COMPLETE   \n",
       "261  NC_021339.1  [PRJNA485481]     COMPLETE   \n",
       "262  NC_024147.1  [PRJNA485481]     COMPLETE   \n",
       "\n",
       "                                                  host  is_annotated  \\\n",
       "0    {'lineage': [{'name': 'cellular organisms', 't...          True   \n",
       "1    {'lineage': [{'name': 'cellular organisms', 't...          True   \n",
       "2    {'lineage': [{'name': 'cellular organisms', 't...          True   \n",
       "3    {'lineage': [{'name': 'cellular organisms', 't...          True   \n",
       "4    {'lineage': [{'name': 'cellular organisms', 't...          True   \n",
       "..                                                 ...           ...   \n",
       "258  {'lineage': [{'name': 'cellular organisms', 't...          True   \n",
       "259  {'lineage': [{'name': 'cellular organisms', 't...          True   \n",
       "260  {'lineage': [{'name': 'cellular organisms', 't...          True   \n",
       "261  {'lineage': [{'name': 'cellular organisms', 't...          True   \n",
       "262  {'lineage': [{'name': 'cellular organisms', 't...          True   \n",
       "\n",
       "    is_lab_host                         lab_host  length  \\\n",
       "0          True          Mycobacterium smegmatis   74483   \n",
       "1           NaN                              NaN   50781   \n",
       "2           NaN                              NaN   71565   \n",
       "3           NaN                              NaN   53396   \n",
       "4          True  Mycobacterium smegmatis mc2 155   69864   \n",
       "..          ...                              ...     ...   \n",
       "258        True  Mycobacterium smegmatis mc2 155   53359   \n",
       "259        True  Mycobacterium smegmatis mc2 155   42395   \n",
       "260        True  Mycobacterium smegmatis mc2 155   52122   \n",
       "261        True            Streptomyces lividans   51077   \n",
       "262        True  Mycobacterium smegmatis mc2 155   57315   \n",
       "\n",
       "                        nucleotide  protein_count          release_date  \\\n",
       "0    {'sequence_hash': '2974DF27'}            142  2006-06-20T00:00:00Z   \n",
       "1     {'sequence_hash': 'CFEA6EA'}             86  2014-03-04T00:00:00Z   \n",
       "2    {'sequence_hash': '52C4C232'}            100  2014-03-04T00:00:00Z   \n",
       "3    {'sequence_hash': '4EC4E167'}             90  2014-03-04T00:00:00Z   \n",
       "4    {'sequence_hash': '4E62B368'}             97  2014-03-11T00:00:00Z   \n",
       "..                             ...            ...                   ...   \n",
       "258  {'sequence_hash': 'A8A0EF4E'}             83  2013-06-03T00:00:00Z   \n",
       "259  {'sequence_hash': '5E5F2E14'}             69  2016-09-27T00:00:00Z   \n",
       "260  {'sequence_hash': '910ADD35'}            101  2013-12-04T00:00:00Z   \n",
       "261  {'sequence_hash': '7D95B565'}             76  2013-06-03T00:00:00Z   \n",
       "262  {'sequence_hash': '7CA1F84F'}             92  2014-05-27T00:00:00Z   \n",
       "\n",
       "    source_database                                          submitter  \\\n",
       "0            RefSeq  {'affiliation': 'National Center for Biotechno...   \n",
       "1            RefSeq  {'affiliation': 'National Center for Biotechno...   \n",
       "2            RefSeq  {'affiliation': 'National Center for Biotechno...   \n",
       "3            RefSeq  {'affiliation': 'National Center for Biotechno...   \n",
       "4            RefSeq  {'affiliation': 'National Center for Biotechno...   \n",
       "..              ...                                                ...   \n",
       "258          RefSeq  {'affiliation': 'National Center for Biotechno...   \n",
       "259          RefSeq  {'affiliation': 'National Center for Biotechno...   \n",
       "260          RefSeq  {'affiliation': 'National Center for Biotechno...   \n",
       "261          RefSeq  {'affiliation': 'National Center for Biotechno...   \n",
       "262          RefSeq  {'affiliation': 'National Center for Biotechno...   \n",
       "\n",
       "              update_date                                              virus  \\\n",
       "0    2023-01-11T00:00:00Z  {'lineage': [{'name': 'Viruses', 'tax_id': 102...   \n",
       "1    2023-01-08T00:00:00Z  {'lineage': [{'name': 'Viruses', 'tax_id': 102...   \n",
       "2    2023-01-08T00:00:00Z  {'lineage': [{'name': 'Viruses', 'tax_id': 102...   \n",
       "3    2023-01-08T00:00:00Z  {'lineage': [{'name': 'Viruses', 'tax_id': 102...   \n",
       "4    2023-01-11T00:00:00Z  {'lineage': [{'name': 'Viruses', 'tax_id': 102...   \n",
       "..                    ...                                                ...   \n",
       "258  2023-01-08T00:00:00Z  {'lineage': [{'name': 'Viruses', 'tax_id': 102...   \n",
       "259  2023-01-09T00:00:00Z  {'lineage': [{'name': 'Viruses', 'tax_id': 102...   \n",
       "260  2023-01-08T00:00:00Z  {'lineage': [{'name': 'Viruses', 'tax_id': 102...   \n",
       "261  2023-01-08T00:00:00Z  {'lineage': [{'name': 'Viruses', 'tax_id': 102...   \n",
       "262  2023-01-08T00:00:00Z  {'lineage': [{'name': 'Viruses', 'tax_id': 102...   \n",
       "\n",
       "                               isolate  \\\n",
       "0                                  NaN   \n",
       "1                                  NaN   \n",
       "2                                  NaN   \n",
       "3                                  NaN   \n",
       "4    {'collection_date': '2009-07-01'}   \n",
       "..                                 ...   \n",
       "258                                NaN   \n",
       "259  {'collection_date': '2014-10-01'}   \n",
       "260  {'collection_date': '2011-09-29'}   \n",
       "261  {'collection_date': '2008-10-30'}   \n",
       "262  {'collection_date': '2012-10-01'}   \n",
       "\n",
       "                                              location  \n",
       "0                                                  NaN  \n",
       "1                                                  NaN  \n",
       "2                                                  NaN  \n",
       "3                                                  NaN  \n",
       "4    {'geographic_location': 'USA: Lafayette, LA', ...  \n",
       "..                                                 ...  \n",
       "258  {'geographic_location': 'China', 'geographic_r...  \n",
       "259  {'geographic_location': 'USA: New Haven, CT', ...  \n",
       "260  {'geographic_location': 'USA: Los Angeles, CA'...  \n",
       "261  {'geographic_location': 'USA: Houston, Texas',...  \n",
       "262  {'geographic_location': 'USA', 'geographic_reg...  \n",
       "\n",
       "[263 rows x 17 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Search on NCBI for extra info\n",
    "import subprocess\n",
    "import json\n",
    "from tqdm import tqdm\n",
    "df_reports = pd.DataFrame()\n",
    "\n",
    "failed_accessions = []\n",
    "\n",
    "for acc in tqdm(PDB_acc['accession'], desc=\"Fetching data from NCBI\", unit=\"acc\"):\n",
    "    #print(acc+\".1\")\n",
    "    res = subprocess.run([datasets, \"summary\", \"virus\", \"genome\", \"accession\", acc+\".1\"],\n",
    "                         capture_output=True,   # capture stdout and stderr\n",
    "                         text=True, \n",
    "                         check=True)          # raise an error if the command fails\n",
    "    # Parse the JSON string\n",
    "    data = json.loads(res.stdout)\n",
    "    \n",
    "    # Check empty data\n",
    "    if \"reports\" not in data:\n",
    "        #print(\"No reports found in the JSON data.\")\n",
    "        failed_accessions.append(acc)\n",
    "        continue\n",
    "\n",
    "    # Convert the 'reports' list to a pandas DataFrame\n",
    "    if df_reports.empty:\n",
    "        df_reports = pd.DataFrame(data['reports'])\n",
    "    else:\n",
    "        df_reports = pd.concat([df_reports, pd.DataFrame(data['reports'])], ignore_index=True)\n",
    "display(df_reports)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5ce2ba0",
   "metadata": {},
   "source": [
    "### Obtaining fasta sequences for each phage\n",
    "Using NCBI CLI:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "532b0ccd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading phage genomes from NCBI: 100%|██████████| 287/287 [03:56<00:00,  1.21acc/s]\n"
     ]
    }
   ],
   "source": [
    "import zipfile\n",
    "import os\n",
    "\n",
    "failed_genomes = []\n",
    "\n",
    "for acc in tqdm(PDB_acc['accession'], desc=\"Downloading phage genomes from NCBI\", unit=\"acc\"):\n",
    "    try:\n",
    "        res = subprocess.run([datasets\n",
    "        , \"download\", \"virus\", \"genome\", \"accession\", acc+\".1\"],\n",
    "                            capture_output=True,   # capture stdout and stderr\n",
    "                            text=True, \n",
    "                            check=True)          # raise an error if the command fails\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        #print(f\"Error fetching accession {acc}: {e.stderr}\")\n",
    "        failed_genomes.append(acc)\n",
    "        continue\n",
    "\n",
    "    phage_name = PDB_acc.loc[PDB_acc['accession'] == acc, 'phage_name'].values[0]\n",
    "\n",
    "    if not os.path.exists(\"ncbi_phage_genomes\"):\n",
    "        os.mkdir(\"ncbi_phage_genomes\")\n",
    "\n",
    "    with zipfile.ZipFile(\"ncbi_dataset.zip\", \"r\") as zip_ref:\n",
    "        file = zip_ref.extract(\"ncbi_dataset/data/genomic.fna\", path=\".\")\n",
    "        with open(file, 'r') as f:\n",
    "            fasta_content = f.read()\n",
    "            with open(f\"ncbi_phage_genomes/{phage_name}.fasta\", \"w\") as output_file:\n",
    "                output_file.write(fasta_content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7867e01e",
   "metadata": {},
   "source": [
    "### Excluding failed accessions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9c35c27e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to download genome for accession: NC_010763\n",
      "Failed to download genome for accession: NC_029018\n",
      "Failed to download genome for accession: NC_022972\n",
      "Failed to download genome for accession: NC_022327\n",
      "Failed to download genome for accession: NC_010762\n",
      "Failed to download genome for accession: NC_021296\n",
      "Failed to download genome for accession: NC_011286\n",
      "Failed to download genome for accession: NC_004682\n",
      "Failed to download genome for accession: NC_011284\n",
      "Failed to download genome for accession: NC_004680\n",
      "Failed to download genome for accession: NC_004686\n",
      "Failed to download genome for accession: NC_014459\n",
      "Failed to download genome for accession: NC_001900\n",
      "Failed to download genome for accession: NC_021306\n",
      "Failed to download genome for accession: NC_009993\n",
      "Failed to download genome for accession: NC_008202\n",
      "Failed to download genome for accession: NC_022054\n",
      "Failed to download genome for accession: NC_004664\n",
      "Failed to download genome for accession: NC_001978\n",
      "Failed to download genome for accession: NC_022326\n",
      "Failed to download genome for accession: NC_011269\n",
      "Failed to download genome for accession: NC_021303\n",
      "Failed to download genome for accession: NC_005345\n",
      "Failed to download genome for accession: NC_008206\n",
      "Failed to fetch summary for accession: NC_010763\n",
      "Failed to fetch summary for accession: NC_029018\n",
      "Failed to fetch summary for accession: NC_022972\n",
      "Failed to fetch summary for accession: NC_022327\n",
      "Failed to fetch summary for accession: NC_010762\n",
      "Failed to fetch summary for accession: NC_021296\n",
      "Failed to fetch summary for accession: NC_011286\n",
      "Failed to fetch summary for accession: NC_004682\n",
      "Failed to fetch summary for accession: NC_011284\n",
      "Failed to fetch summary for accession: NC_004680\n",
      "Failed to fetch summary for accession: NC_004686\n",
      "Failed to fetch summary for accession: NC_014459\n",
      "Failed to fetch summary for accession: NC_001900\n",
      "Failed to fetch summary for accession: NC_021306\n",
      "Failed to fetch summary for accession: NC_009993\n",
      "Failed to fetch summary for accession: NC_008202\n",
      "Failed to fetch summary for accession: NC_022054\n",
      "Failed to fetch summary for accession: NC_004664\n",
      "Failed to fetch summary for accession: NC_001978\n",
      "Failed to fetch summary for accession: NC_022326\n",
      "Failed to fetch summary for accession: NC_011269\n",
      "Failed to fetch summary for accession: NC_021303\n",
      "Failed to fetch summary for accession: NC_005345\n",
      "Failed to fetch summary for accession: NC_008206\n",
      "Filtering\n"
     ]
    },
    {
     "ename": "OSError",
     "evalue": "Cannot save file into a non-existent directory: 'data_prod'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[19], line 12\u001b[0m\n\u001b[0;32m      9\u001b[0m PDB_acc_success \u001b[38;5;241m=\u001b[39m PDB_acc[\u001b[38;5;241m~\u001b[39mPDB_acc[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124maccession\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39misin(failed_accessions \u001b[38;5;241m+\u001b[39m failed_genomes)]\n\u001b[0;32m     10\u001b[0m df_reports_success \u001b[38;5;241m=\u001b[39m df_reports[\u001b[38;5;241m~\u001b[39mdf_reports[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124maccession\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39misin(failed_accessions \u001b[38;5;241m+\u001b[39m failed_genomes)]\n\u001b[1;32m---> 12\u001b[0m \u001b[43mPDB_acc_success\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdata_prod/PDB_acc.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[0;32m     13\u001b[0m df_reports_success\u001b[38;5;241m.\u001b[39mto_csv(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdata_prod/NCBI_virus_genome_summary_PDB_phages.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m, index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "File \u001b[1;32mc:\\Users\\asbjo\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\pandas\\util\\_decorators.py:333\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    327\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[0;32m    328\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[0;32m    329\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39m_format_argument_list(allow_args)),\n\u001b[0;32m    330\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[0;32m    331\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mfind_stack_level(),\n\u001b[0;32m    332\u001b[0m     )\n\u001b[1;32m--> 333\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m func(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\asbjo\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\pandas\\core\\generic.py:3986\u001b[0m, in \u001b[0;36mNDFrame.to_csv\u001b[1;34m(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, lineterminator, chunksize, date_format, doublequote, escapechar, decimal, errors, storage_options)\u001b[0m\n\u001b[0;32m   3975\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m, ABCDataFrame) \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mto_frame()\n\u001b[0;32m   3977\u001b[0m formatter \u001b[38;5;241m=\u001b[39m DataFrameFormatter(\n\u001b[0;32m   3978\u001b[0m     frame\u001b[38;5;241m=\u001b[39mdf,\n\u001b[0;32m   3979\u001b[0m     header\u001b[38;5;241m=\u001b[39mheader,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   3983\u001b[0m     decimal\u001b[38;5;241m=\u001b[39mdecimal,\n\u001b[0;32m   3984\u001b[0m )\n\u001b[1;32m-> 3986\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mDataFrameRenderer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mformatter\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m   3987\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpath_or_buf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3988\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlineterminator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlineterminator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3989\u001b[0m \u001b[43m    \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msep\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3990\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3991\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3992\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3993\u001b[0m \u001b[43m    \u001b[49m\u001b[43mquoting\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mquoting\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3994\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3995\u001b[0m \u001b[43m    \u001b[49m\u001b[43mindex_label\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindex_label\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3996\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3997\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunksize\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunksize\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3998\u001b[0m \u001b[43m    \u001b[49m\u001b[43mquotechar\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mquotechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3999\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdate_format\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdate_format\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   4000\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdoublequote\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdoublequote\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   4001\u001b[0m \u001b[43m    \u001b[49m\u001b[43mescapechar\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mescapechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   4002\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   4003\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\asbjo\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\pandas\\io\\formats\\format.py:1014\u001b[0m, in \u001b[0;36mDataFrameRenderer.to_csv\u001b[1;34m(self, path_or_buf, encoding, sep, columns, index_label, mode, compression, quoting, quotechar, lineterminator, chunksize, date_format, doublequote, escapechar, errors, storage_options)\u001b[0m\n\u001b[0;32m    993\u001b[0m     created_buffer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[0;32m    995\u001b[0m csv_formatter \u001b[38;5;241m=\u001b[39m CSVFormatter(\n\u001b[0;32m    996\u001b[0m     path_or_buf\u001b[38;5;241m=\u001b[39mpath_or_buf,\n\u001b[0;32m    997\u001b[0m     lineterminator\u001b[38;5;241m=\u001b[39mlineterminator,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1012\u001b[0m     formatter\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfmt,\n\u001b[0;32m   1013\u001b[0m )\n\u001b[1;32m-> 1014\u001b[0m \u001b[43mcsv_formatter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msave\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1016\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m created_buffer:\n\u001b[0;32m   1017\u001b[0m     \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(path_or_buf, StringIO)\n",
      "File \u001b[1;32mc:\\Users\\asbjo\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\pandas\\io\\formats\\csvs.py:251\u001b[0m, in \u001b[0;36mCSVFormatter.save\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    247\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    248\u001b[0m \u001b[38;5;124;03mCreate the writer & save.\u001b[39;00m\n\u001b[0;32m    249\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    250\u001b[0m \u001b[38;5;66;03m# apply compression and byte/text conversion\u001b[39;00m\n\u001b[1;32m--> 251\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    252\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    253\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    254\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    255\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    256\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    257\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    258\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m handles:\n\u001b[0;32m    259\u001b[0m     \u001b[38;5;66;03m# Note: self.encoding is irrelevant here\u001b[39;00m\n\u001b[0;32m    260\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwriter \u001b[38;5;241m=\u001b[39m csvlib\u001b[38;5;241m.\u001b[39mwriter(\n\u001b[0;32m    261\u001b[0m         handles\u001b[38;5;241m.\u001b[39mhandle,\n\u001b[0;32m    262\u001b[0m         lineterminator\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlineterminator,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    267\u001b[0m         quotechar\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mquotechar,\n\u001b[0;32m    268\u001b[0m     )\n\u001b[0;32m    270\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_save()\n",
      "File \u001b[1;32mc:\\Users\\asbjo\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\pandas\\io\\common.py:749\u001b[0m, in \u001b[0;36mget_handle\u001b[1;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[0;32m    747\u001b[0m \u001b[38;5;66;03m# Only for write methods\u001b[39;00m\n\u001b[0;32m    748\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode \u001b[38;5;129;01mand\u001b[39;00m is_path:\n\u001b[1;32m--> 749\u001b[0m     \u001b[43mcheck_parent_directory\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    751\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m compression:\n\u001b[0;32m    752\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m compression \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mzstd\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m    753\u001b[0m         \u001b[38;5;66;03m# compression libraries do not like an explicit text-mode\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\asbjo\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\pandas\\io\\common.py:616\u001b[0m, in \u001b[0;36mcheck_parent_directory\u001b[1;34m(path)\u001b[0m\n\u001b[0;32m    614\u001b[0m parent \u001b[38;5;241m=\u001b[39m Path(path)\u001b[38;5;241m.\u001b[39mparent\n\u001b[0;32m    615\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m parent\u001b[38;5;241m.\u001b[39mis_dir():\n\u001b[1;32m--> 616\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m(\u001b[38;5;124mrf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot save file into a non-existent directory: \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mparent\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[1;31mOSError\u001b[0m: Cannot save file into a non-existent directory: 'data_prod'"
     ]
    }
   ],
   "source": [
    "for acc in failed_genomes:\n",
    "    print(f\"Failed to download genome for accession: {acc}\")\n",
    "\n",
    "for acc in failed_accessions:\n",
    "    print(f\"Failed to fetch summary for accession: {acc}\")\n",
    "\n",
    "print(\"Filtering\")\n",
    "\n",
    "PDB_acc_success = PDB_acc[~PDB_acc['accession'].isin(failed_accessions + failed_genomes)]\n",
    "df_reports_success = df_reports[~df_reports['accession'].isin(failed_accessions + failed_genomes)]\n",
    "\n",
    "PDB_acc_success.to_csv(\"data_prod/PDB_acc.csv\", sep=\"\\t\", index=False)\n",
    "df_reports_success.to_csv(\"data_prod/NCBI_virus_genome_summary_PDB_phages.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e67c312",
   "metadata": {},
   "source": [
    "### Phages against clusters\n",
    "Clusters identify which bacteria each phage can infect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27e4e568",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phage Name</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Subcluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20ES</td>\n",
       "      <td>A</td>\n",
       "      <td>A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>244</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>32HC</td>\n",
       "      <td>Z</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>39HC</td>\n",
       "      <td>B</td>\n",
       "      <td>B6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40AC</td>\n",
       "      <td>A</td>\n",
       "      <td>A17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5573</th>\n",
       "      <td>Zorro</td>\n",
       "      <td>AK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5574</th>\n",
       "      <td>Zucker</td>\n",
       "      <td>FN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5575</th>\n",
       "      <td>Zuko</td>\n",
       "      <td>BR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5576</th>\n",
       "      <td>Zulu</td>\n",
       "      <td>A</td>\n",
       "      <td>A6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5577</th>\n",
       "      <td>ZygoTaiga</td>\n",
       "      <td>C</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5578 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Phage Name Cluster Subcluster\n",
       "0          20ES       A         A2\n",
       "1           244       E        NaN\n",
       "2          32HC       Z        NaN\n",
       "3          39HC       B         B6\n",
       "4          40AC       A        A17\n",
       "...         ...     ...        ...\n",
       "5573      Zorro      AK        NaN\n",
       "5574     Zucker      FN        NaN\n",
       "5575       Zuko      BR        NaN\n",
       "5576       Zulu       A         A6\n",
       "5577  ZygoTaiga       C         C1\n",
       "\n",
       "[5578 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PDB_phage_cluster = pd.read_csv(\"PhagesDB_Data.txt\", sep=\"\\t\", header=0)\n",
    "PDB_phage_cluster"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
