{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "42c7683a",
   "metadata": {},
   "source": [
    "# Neural Network for Kmers\n",
    "Training neural network not on minhashes, but om kmers. Perhaps too large for bacteria genomes, but should be feasible for phage genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2f7a6b1",
   "metadata": {},
   "source": [
    "## Prepping data \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "60ab48b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python39.zip', '/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9', '/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/lib-dynload', '', '/Users/asbjornhansen/Library/Python/3.9/lib/python/site-packages', '/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/site-packages', '../scripts/', '../scripts/']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import os, sys\n",
    "from tqdm import tqdm\n",
    "\n",
    "K = 12\n",
    "raw_data_path = \"../raw_data/\"\n",
    "data_prod_path = \"../data_prod/\"\n",
    "scripts_path = \"../scripts/\"\n",
    "\n",
    "sys.path.append(scripts_path)\n",
    "print(sys.path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bee3dc7f",
   "metadata": {},
   "source": [
    "#### Kmer phage genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fad735b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total sequences to process: 23\n",
      "Converting 23 sequences to 12-mer frequency DataFrame...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing sequences: 100%|██████████| 23/23 [00:00<00:00, 67.16seq/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Combining k-mer counts into DataFrame...\n",
      "Generated k-mer DataFrame with shape: (23, 886272)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CGCAGGGTGTCT</th>\n",
       "      <th>GGTTCGAATCCT</th>\n",
       "      <th>AGGTTCGAATCC</th>\n",
       "      <th>GCAGGGTGTCTT</th>\n",
       "      <th>TATTCGCGTGTT</th>\n",
       "      <th>GTTTCCTGCAAC</th>\n",
       "      <th>TGTTTCCTGCAA</th>\n",
       "      <th>GTGTTTCCTGCA</th>\n",
       "      <th>CGTGTTTCCTGC</th>\n",
       "      <th>GCGTGTTTCCTG</th>\n",
       "      <th>...</th>\n",
       "      <th>TGAGAGGCTTTG</th>\n",
       "      <th>GAGAGGCTTTGA</th>\n",
       "      <th>AGAGGCTTTGAG</th>\n",
       "      <th>GAGGCTTTGAGT</th>\n",
       "      <th>AGGCTTTGAGTT</th>\n",
       "      <th>GGCTTTGAGTTG</th>\n",
       "      <th>CTTTGAGTTGAC</th>\n",
       "      <th>CGCTGAGAGGCT</th>\n",
       "      <th>TTTGAGTTGACA</th>\n",
       "      <th>GCTGAGAGGCTT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lelliottia_phage_Pantea</th>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lelliottia_phage_Rap</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lelliottia_phage_Zann</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Morganella_phage_Rip</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Morganella_phage_Rup</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Morganella_phage_Slaad</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Crus</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Gander</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Guf</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Hoejben</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Magnum</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Vims</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Serratia_phage_Echoes</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Serratia_phage_Galvinrad</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Serratia_phage_Uther</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Mimer</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Abuela</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Koroua</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Sabo</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Taid</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Ymer</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Amona</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium_phage_Poppous</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23 rows × 886272 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              CGCAGGGTGTCT  GGTTCGAATCCT  AGGTTCGAATCC  \\\n",
       "Lelliottia_phage_Pantea           0.000033      0.000033      0.000033   \n",
       "Lelliottia_phage_Rap              0.000000      0.000000      0.000000   \n",
       "Lelliottia_phage_Zann             0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Rip              0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Rup              0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Slaad            0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Crus         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Gander       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Guf          0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Hoejben      0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Magnum       0.000000      0.000042      0.000032   \n",
       "Pectobacterium_phage_Vims         0.000000      0.000021      0.000021   \n",
       "Serratia_phage_Echoes             0.000000      0.000000      0.000000   \n",
       "Serratia_phage_Galvinrad          0.000000      0.000000      0.000000   \n",
       "Serratia_phage_Uther              0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Mimer        0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Abuela       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Koroua       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Sabo         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Taid         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Ymer         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Amona        0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Poppous      0.000000      0.000000      0.000000   \n",
       "\n",
       "                              GCAGGGTGTCTT  TATTCGCGTGTT  GTTTCCTGCAAC  \\\n",
       "Lelliottia_phage_Pantea           0.000027      0.000027      0.000027   \n",
       "Lelliottia_phage_Rap              0.000000      0.000000      0.000000   \n",
       "Lelliottia_phage_Zann             0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Rip              0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Rup              0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Slaad            0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Crus         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Gander       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Guf          0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Hoejben      0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Magnum       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Vims         0.000000      0.000021      0.000000   \n",
       "Serratia_phage_Echoes             0.000000      0.000000      0.000000   \n",
       "Serratia_phage_Galvinrad          0.000000      0.000000      0.000000   \n",
       "Serratia_phage_Uther              0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Mimer        0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Abuela       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Koroua       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Sabo         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Taid         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Ymer         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Amona        0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Poppous      0.000000      0.000000      0.000000   \n",
       "\n",
       "                              TGTTTCCTGCAA  GTGTTTCCTGCA  CGTGTTTCCTGC  \\\n",
       "Lelliottia_phage_Pantea           0.000027      0.000027      0.000027   \n",
       "Lelliottia_phage_Rap              0.000000      0.000000      0.000000   \n",
       "Lelliottia_phage_Zann             0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Rip              0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Rup              0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Slaad            0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Crus         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Gander       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Guf          0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Hoejben      0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Magnum       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Vims         0.000000      0.000000      0.000000   \n",
       "Serratia_phage_Echoes             0.000000      0.000000      0.000000   \n",
       "Serratia_phage_Galvinrad          0.000000      0.000000      0.000000   \n",
       "Serratia_phage_Uther              0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Mimer        0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Abuela       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Koroua       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Sabo         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Taid         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Ymer         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Amona        0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Poppous      0.000000      0.000000      0.000000   \n",
       "\n",
       "                              GCGTGTTTCCTG  ...  TGAGAGGCTTTG  GAGAGGCTTTGA  \\\n",
       "Lelliottia_phage_Pantea           0.000027  ...      0.000000      0.000000   \n",
       "Lelliottia_phage_Rap              0.000000  ...      0.000000      0.000000   \n",
       "Lelliottia_phage_Zann             0.000000  ...      0.000000      0.000000   \n",
       "Morganella_phage_Rip              0.000000  ...      0.000000      0.000000   \n",
       "Morganella_phage_Rup              0.000000  ...      0.000000      0.000000   \n",
       "Morganella_phage_Slaad            0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Crus         0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Gander       0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Guf          0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Hoejben      0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Magnum       0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Vims         0.000000  ...      0.000000      0.000000   \n",
       "Serratia_phage_Echoes             0.000000  ...      0.000000      0.000000   \n",
       "Serratia_phage_Galvinrad          0.000000  ...      0.000000      0.000000   \n",
       "Serratia_phage_Uther              0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Mimer        0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Abuela       0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Koroua       0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Sabo         0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Taid         0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Ymer         0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Amona        0.000000  ...      0.000000      0.000000   \n",
       "Pectobacterium_phage_Poppous      0.000000  ...      0.000024      0.000024   \n",
       "\n",
       "                              AGAGGCTTTGAG  GAGGCTTTGAGT  AGGCTTTGAGTT  \\\n",
       "Lelliottia_phage_Pantea           0.000000      0.000000      0.000000   \n",
       "Lelliottia_phage_Rap              0.000000      0.000000      0.000000   \n",
       "Lelliottia_phage_Zann             0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Rip              0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Rup              0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Slaad            0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Crus         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Gander       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Guf          0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Hoejben      0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Magnum       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Vims         0.000000      0.000000      0.000000   \n",
       "Serratia_phage_Echoes             0.000000      0.000000      0.000000   \n",
       "Serratia_phage_Galvinrad          0.000000      0.000000      0.000000   \n",
       "Serratia_phage_Uther              0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Mimer        0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Abuela       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Koroua       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Sabo         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Taid         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Ymer         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Amona        0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Poppous      0.000024      0.000024      0.000024   \n",
       "\n",
       "                              GGCTTTGAGTTG  CTTTGAGTTGAC  CGCTGAGAGGCT  \\\n",
       "Lelliottia_phage_Pantea           0.000000      0.000000      0.000000   \n",
       "Lelliottia_phage_Rap              0.000000      0.000000      0.000000   \n",
       "Lelliottia_phage_Zann             0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Rip              0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Rup              0.000000      0.000000      0.000000   \n",
       "Morganella_phage_Slaad            0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Crus         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Gander       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Guf          0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Hoejben      0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Magnum       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Vims         0.000000      0.000000      0.000000   \n",
       "Serratia_phage_Echoes             0.000000      0.000000      0.000000   \n",
       "Serratia_phage_Galvinrad          0.000000      0.000000      0.000000   \n",
       "Serratia_phage_Uther              0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Mimer        0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Abuela       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Koroua       0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Sabo         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Taid         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Ymer         0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Amona        0.000000      0.000000      0.000000   \n",
       "Pectobacterium_phage_Poppous      0.000024      0.000024      0.000024   \n",
       "\n",
       "                              TTTGAGTTGACA  GCTGAGAGGCTT  \n",
       "Lelliottia_phage_Pantea           0.000000      0.000000  \n",
       "Lelliottia_phage_Rap              0.000000      0.000000  \n",
       "Lelliottia_phage_Zann             0.000000      0.000000  \n",
       "Morganella_phage_Rip              0.000000      0.000000  \n",
       "Morganella_phage_Rup              0.000000      0.000000  \n",
       "Morganella_phage_Slaad            0.000000      0.000000  \n",
       "Pectobacterium_phage_Crus         0.000000      0.000000  \n",
       "Pectobacterium_phage_Gander       0.000000      0.000000  \n",
       "Pectobacterium_phage_Guf          0.000000      0.000000  \n",
       "Pectobacterium_phage_Hoejben      0.000000      0.000000  \n",
       "Pectobacterium_phage_Magnum       0.000000      0.000000  \n",
       "Pectobacterium_phage_Vims         0.000000      0.000000  \n",
       "Serratia_phage_Echoes             0.000000      0.000000  \n",
       "Serratia_phage_Galvinrad          0.000000      0.000000  \n",
       "Serratia_phage_Uther              0.000000      0.000000  \n",
       "Pectobacterium_phage_Mimer        0.000000      0.000000  \n",
       "Pectobacterium_phage_Abuela       0.000000      0.000000  \n",
       "Pectobacterium_phage_Koroua       0.000000      0.000000  \n",
       "Pectobacterium_phage_Sabo         0.000000      0.000000  \n",
       "Pectobacterium_phage_Taid         0.000000      0.000000  \n",
       "Pectobacterium_phage_Ymer         0.000000      0.000000  \n",
       "Pectobacterium_phage_Amona        0.000000      0.000000  \n",
       "Pectobacterium_phage_Poppous      0.000024      0.000024  \n",
       "\n",
       "[23 rows x 886272 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from manipulations import fasta_to_kmerdf\n",
    "all_phage_kmers = fasta_to_kmerdf(raw_data_path+\"phagehost_KU/phage_cleaned.fasta\", k=K)\n",
    "display(all_phage_kmers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8806f11c",
   "metadata": {},
   "source": [
    "#### Hostrange data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c39429bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bacteria lookup dictionary created with 110 entries.\n",
      "{'J14_21_reoriented': 'Acinetobacter calcoaceticus', 'J53_21_reoriented': 'Acinetobacter calcoaceticus', 'J105_22_reoriented': 'Chishuiella', 'J46_21_reoriented': 'Chryseobacterium', 'J50_21_reoriented': 'Chryseobacterium', 'J2264_1_22_KMC_reoriented': 'Chryseobacterium', 'J2264_3_22_KMC_reoriented': 'Chryseobacterium', 'J63_22_reoriented': 'Chryseobacterium', 'J64_22_reoriented': 'Chryseobacterium', 'J1_21_reoriented': 'Lelliottia', 'J91_22_reoriented': 'Lelliottia', 'J51_21_reoriented': 'Morganella morganii', 'J57_21_reoriented': 'Morganella morganii', 'J10_21_reoriented': 'Pectobacterium atrosepticum', 'J11_21_reoriented': 'Pectobacterium atrosepticum', 'J126_23_reoriented': 'Pectobacterium atrosepticum', 'J12_21_reoriented': 'Pectobacterium atrosepticum', 'J16_21_reoriented': 'Pectobacterium atrosepticum', 'J22_21_reoriented': 'Pectobacterium atrosepticum', 'J28_21_reoriented': 'Pectobacterium atrosepticum', 'J33_21_reoriented': 'Pectobacterium atrosepticum', 'J38_21_reoriented': 'Pectobacterium atrosepticum', 'J39_21_reoriented': 'Pectobacterium atrosepticum', 'J40_21_reoriented': 'Pectobacterium atrosepticum', 'J56_21_reoriented': 'Pectobacterium atrosepticum', 'J6_21_reoriented': 'Pectobacterium atrosepticum', 'J7_21_reoriented': 'Pectobacterium atrosepticum', 'J8_21_reoriented': 'Pectobacterium atrosepticum', 'J116_23_reoriented': 'Pectobacterium atrosepticum', 'J118_23_reoriented': 'Pectobacterium atrosepticum', 'J121_23_reoriented': 'Pectobacterium atrosepticum', 'J125_23_reoriented': 'Pectobacterium atrosepticum', 'J70_21_reoriented': 'Pectobacterium atrosepticum', 'J78_22_reoriented': 'Pectobacterium atrosepticum', 'J81_22_reoriented': 'Pectobacterium atrosepticum', 'J82_22_reoriented': 'Pectobacterium atrosepticum', 'J84_22_reoriented': 'Pectobacterium atrosepticum', 'J119_23_reoriented': 'Pectobacterium brasiliense', 'J127_23_reoriented': 'Pectobacterium brasiliense', 'J13_21_reoriented': 'Pectobacterium brasiliense', 'J20_21_reoriented': 'Pectobacterium brasiliense', 'J21_21_reoriented': 'Pectobacterium brasiliense', 'J24_21_reoriented': 'Pectobacterium brasiliense', 'J25_21_reoriented': 'Pectobacterium brasiliense', 'J26_21_reoriented': 'Pectobacterium brasiliense', 'J31_21_reoriented': 'Pectobacterium brasiliense', 'J32_21_reoriented': 'Pectobacterium brasiliense', 'J34_21_reoriented': 'Pectobacterium brasiliense', 'J36_21_reoriented': 'Pectobacterium brasiliense', 'J37_21_reoriented': 'Pectobacterium brasiliense', 'J41_21_reoriented': 'Pectobacterium brasiliense', 'J44_21_reoriented': 'Pectobacterium brasiliense', 'J45_21_reoriented': 'Pectobacterium brasiliense', 'J47_21_reoriented': 'Pectobacterium brasiliense', 'J52_21_reoriented': 'Pectobacterium brasiliense', 'J59_21_reoriented': 'Pectobacterium brasiliense', 'J115_23_reoriented': 'Pectobacterium brasiliense', 'J117_23_reoriented': 'Pectobacterium brasiliense', 'J128_23_reoriented': 'Pectobacterium brasiliense', 'J134_23_reoriented': 'Pectobacterium brasiliense', 'J3009LEM_22_KMC_reoriented': 'Pectobacterium brasiliense', 'J3009_2_22_KMC_reoriented': 'Pectobacterium brasiliense', 'J69_21_reoriented': 'Pectobacterium brasiliense', 'J73_21_reoriented': 'Pectobacterium brasiliense', 'J77_22_reoriented': 'Pectobacterium brasiliense', 'J86_22_reoriented': 'Pectobacterium brasiliense', 'J97_22_reoriented': 'Pectobacterium brasiliense', 'J98_22_reoriented': 'Pectobacterium brasiliense', 'J54_21_reoriented': 'Pectobacterium carotovorum', 'J100_22_reoriented': 'Pectobacterium carotovorum', 'J102_22_reoriented': 'Pectobacterium carotovorum', 'J74_21_reoriented': 'Pectobacterium carotovorum', 'J80_22_reoriented': 'Pectobacterium carotovorum', 'J27_21_reoriented': 'Pectobacterium parmentieri', 'J29_21_reoriented': 'Pectobacterium parmentieri', 'J30_21_reoriented': 'Pectobacterium parmentieri', 'J42_21_reoriented': 'Pectobacterium parmentieri', 'J49_21_reoriented': 'Pectobacterium parmentieri', 'FO3A_23_KMC_reoriented': 'Pectobacterium parmentieri', 'J103_22_reoriented': 'Pectobacterium parmentieri', 'J104_22_reoriented': 'Pectobacterium parmentieri', 'J106_22_reoriented': 'Pectobacterium parmentieri', 'J108_22_reoriented': 'Pectobacterium parmentieri', 'J62_22_reoriented': 'Pectobacterium parmentieri', 'J65_21_reoriented': 'Pectobacterium parmentieri', 'J68_21_reoriented': 'Pectobacterium parmentieri', 'J83_22_reoriented': 'Pectobacterium parmentieri', 'J93_22_reoriented': 'Pectobacterium parmentieri', 'J95_22_reoriented': 'Pectobacterium parmentieri', 'J96_22_reoriented': 'Pectobacterium parmentieri', 'J17_21_reoriented': 'Pectobacterium polaris', 'J18_21_reoriented': 'Pectobacterium polaris', 'J19_21_reoriented': 'Pectobacterium polaris', 'J48_21_reoriented': 'Pectobacterium polaris', 'J3009_1_22_KMC_reoriented': 'Pectobacterium polaris', 'J60_21_reoriented': 'Pectobacterium polaris', 'J61_21_reoriented': 'Pectobacterium polaris', 'J66_21_reoriented': 'Pectobacterium polaris', 'J75_21_reoriented': 'Pectobacterium polaris', 'J76_21_reoriented': 'Pectobacterium polaris', 'J79_22_reoriented': 'Pectobacterium polaris', 'J94_22_reoriented': 'Pectobacterium polaris', 'J99_22_reoriented': 'Pectobacterium polaris', 'J2_21_reoriented': 'Pectobacterium punjabense', 'J3_21_reoriented': 'Pectobacterium punjabense', 'J109_23_reoriented': 'Pseudomonas chlororaphis', 'J101_22_reoriented': 'Pseudomonas marginalis', 'J15_21_reoriented': 'Serratia liquefaciens', 'J4_21_reoriented': 'Serratia plymuthica', 'J5_21_reoriented': 'Vagococcus'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phage</th>\n",
       "      <th>Ymer</th>\n",
       "      <th>Taid</th>\n",
       "      <th>Poppous</th>\n",
       "      <th>Koroua</th>\n",
       "      <th>Abuela</th>\n",
       "      <th>Amona</th>\n",
       "      <th>Sabo</th>\n",
       "      <th>Mimer</th>\n",
       "      <th>Crus</th>\n",
       "      <th>...</th>\n",
       "      <th>Vims</th>\n",
       "      <th>Echoes</th>\n",
       "      <th>Galvinrad</th>\n",
       "      <th>Uther</th>\n",
       "      <th>Rip</th>\n",
       "      <th>Rup</th>\n",
       "      <th>Slaad</th>\n",
       "      <th>Pantea</th>\n",
       "      <th>Rap</th>\n",
       "      <th>Zann</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>J14_21_reoriented</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>J53_21_reoriented</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>J105_22_reoriented</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>J46_21_reoriented</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>J50_21_reoriented</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                phage Ymer Taid Poppous Koroua Abuela Amona Sabo Mimer  Crus  \\\n",
       "0   J14_21_reoriented  NaN  NaN     NaN    NaN    NaN   NaN  NaN   NaN   NaN   \n",
       "1   J53_21_reoriented  NaN  NaN     NaN    NaN    NaN   NaN  NaN   NaN   NaN   \n",
       "2  J105_22_reoriented  NaN  NaN     NaN    NaN    NaN   NaN  NaN   NaN   NaN   \n",
       "3   J46_21_reoriented  NaN  NaN     NaN    NaN    NaN   NaN  NaN   NaN   NaN   \n",
       "4   J50_21_reoriented  NaN  NaN     NaN    NaN    NaN   NaN  NaN   NaN   NaN   \n",
       "\n",
       "   ... Vims Echoes Galvinrad Uther  Rip  Rup Slaad Pantea  Rap Zann  \n",
       "0  ...  NaN    NaN       NaN   NaN  NaN  NaN   NaN    NaN  NaN  NaN  \n",
       "1  ...  NaN    NaN       NaN   NaN  NaN  NaN   NaN    NaN  NaN  NaN  \n",
       "2  ...  NaN    NaN       NaN   NaN  NaN  NaN   NaN    NaN  NaN  NaN  \n",
       "3  ...  NaN    NaN       NaN   NaN  NaN  NaN   NaN    NaN  NaN  NaN  \n",
       "4  ...  NaN    NaN       NaN   NaN  NaN  NaN   NaN    NaN  NaN  NaN  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "{'Ymer': nan,\n",
       " 'Taid': nan,\n",
       " 'Poppous': nan,\n",
       " 'Koroua': nan,\n",
       " 'Abuela': nan,\n",
       " 'Amona': nan,\n",
       " 'Sabo': nan,\n",
       " 'Mimer': nan,\n",
       " 'Crus': nan,\n",
       " 'Gander': nan,\n",
       " 'Guf': nan,\n",
       " 'Hoejben': nan,\n",
       " 'Magnum': 200000000,\n",
       " 'Vims': nan,\n",
       " 'Echoes': nan,\n",
       " 'Galvinrad': nan,\n",
       " 'Uther': nan,\n",
       " 'Rip': nan,\n",
       " 'Rup': nan,\n",
       " 'Slaad': nan,\n",
       " 'Pantea': nan,\n",
       " 'Rap': nan,\n",
       " 'Zann': nan}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from manipulations import binarize_host_range, short_species_name\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# Load the host range data from the Excel file\n",
    "file_path = raw_data_path + \"phagehost_KU/Hostrange_data_all_crisp_iso.xlsx\"\n",
    "sheet_name = \"sum_hostrange\"  # replace with your sheet name\n",
    "host_range_df = pd.read_excel(\n",
    "    file_path,\n",
    "    sheet_name='sum_hostrange',\n",
    "    header=1).drop(columns=[\"isolate ID\", \"Hostrange_analysis\", \"Phage\"])\n",
    "\n",
    "# Create a lookup dictionary for bacteria species based on Seq ID - dict\n",
    "bact_lookup = host_range_df[[\"Seq ID\", \"Species\"]].drop_duplicates(subset=['Seq ID']).set_index('Seq ID').to_dict()['Species']\n",
    "print(\"Bacteria lookup dictionary created with\", len(bact_lookup), \"entries.\")\n",
    "print(bact_lookup)\n",
    "\n",
    "# Make Seq ID to phage name mapping - pandas df\n",
    "host_range_df = host_range_df.drop(columns=[\"Species\"]).set_index('Seq ID').rename_axis('phage').reset_index()\n",
    "display(host_range_df.head())\n",
    "\n",
    "# Convert the host range data into a dictionary\n",
    "host_range_data = {}\n",
    "for index, row in host_range_df.iterrows():\n",
    "    cleaned_index = row[1:].index.str.replace(\" \", \"\")\n",
    "    curr_bact_series = row[1:]\n",
    "    curr_bact_series.index = cleaned_index\n",
    "    host_range_data[row['phage']] = curr_bact_series.to_dict()\n",
    "\n",
    "host_range_data[\"J10_21_reoriented\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b39b748e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Processing host range data for bacteria species: Pseudomonas marginalis\n",
      "Seq IDs for selected species: ['J101_22_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 0, 'Taid': 0, 'Poppous': 0, 'Koroua': 0, 'Abuela': 0, 'Amona': 0, 'Sabo': 0, 'Mimer': 0, 'Crus': 0, 'Gander': 0, 'Guf': 0, 'Hoejben': 0, 'Magnum': 0, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 0, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 0, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Pantea_P.marginalis...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Rap_P.marginalis...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Zann_P.marginalis...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Rip_P.marginalis...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Rup_P.marginalis...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Slaad_P.marginalis...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Crus_P.marginalis...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Gander_P.marginalis...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Guf_P.marginalis...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Hoejben_P.marginalis...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Magnum_P.marginalis...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Vims_P.marginalis...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Echoes_P.marginalis...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Galvinrad_P.marginalis...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Uther_P.marginalis...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Mimer_P.marginalis...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Abuela_P.marginalis...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Koroua_P.marginalis...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Sabo_P.marginalis...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Taid_P.marginalis...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Ymer_P.marginalis...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Amona_P.marginalis...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_marginalis/+Poppous_P.marginalis...\n",
      "\n",
      "Processing host range data for bacteria species: Serratia liquefaciens\n",
      "Seq IDs for selected species: ['J15_21_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 0, 'Taid': 0, 'Poppous': 0, 'Koroua': 0, 'Abuela': 0, 'Amona': 0, 'Sabo': 0, 'Mimer': 0, 'Crus': 0, 'Gander': 0, 'Guf': 0, 'Hoejben': 0, 'Magnum': 0, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 1, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 0, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Pantea_S.liquefaciens...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Rap_S.liquefaciens...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Zann_S.liquefaciens...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Rip_S.liquefaciens...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Rup_S.liquefaciens...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Slaad_S.liquefaciens...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Crus_S.liquefaciens...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Gander_S.liquefaciens...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Guf_S.liquefaciens...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Hoejben_S.liquefaciens...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Magnum_S.liquefaciens...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Vims_S.liquefaciens...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Echoes_S.liquefaciens...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Galvinrad_S.liquefaciens...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Uther_S.liquefaciens...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Mimer_S.liquefaciens...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Abuela_S.liquefaciens...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Koroua_S.liquefaciens...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Sabo_S.liquefaciens...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Taid_S.liquefaciens...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Ymer_S.liquefaciens...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Amona_S.liquefaciens...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_liquefaciens/+Poppous_S.liquefaciens...\n",
      "\n",
      "Processing host range data for bacteria species: Lelliottia\n",
      "Seq IDs for selected species: ['J1_21_reoriented', 'J91_22_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 0, 'Taid': 0, 'Poppous': 0, 'Koroua': 0, 'Abuela': 0, 'Amona': 0, 'Sabo': 0, 'Mimer': 0, 'Crus': 0, 'Gander': 0, 'Guf': 0, 'Hoejben': 0, 'Magnum': 0, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 0, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 1, 'Zann': 1}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Pantea_Lelliottia...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Rap_Lelliottia...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Zann_Lelliottia...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Rip_Lelliottia...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Rup_Lelliottia...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Slaad_Lelliottia...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Crus_Lelliottia...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Gander_Lelliottia...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Guf_Lelliottia...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Hoejben_Lelliottia...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Magnum_Lelliottia...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Vims_Lelliottia...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Echoes_Lelliottia...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Galvinrad_Lelliottia...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Uther_Lelliottia...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Mimer_Lelliottia...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Abuela_Lelliottia...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Koroua_Lelliottia...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Sabo_Lelliottia...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Taid_Lelliottia...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Ymer_Lelliottia...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Amona_Lelliottia...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Lelliottia/+Poppous_Lelliottia...\n",
      "\n",
      "Processing host range data for bacteria species: Pectobacterium carotovorum\n",
      "Seq IDs for selected species: ['J54_21_reoriented', 'J100_22_reoriented', 'J102_22_reoriented', 'J74_21_reoriented', 'J80_22_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 1, 'Taid': 0, 'Poppous': 0, 'Koroua': 1, 'Abuela': 1, 'Amona': 1, 'Sabo': 0, 'Mimer': 1, 'Crus': 1, 'Gander': 1, 'Guf': 1, 'Hoejben': 1, 'Magnum': 0, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 0, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 0, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Pantea_P.carotovorum...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Rap_P.carotovorum...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Zann_P.carotovorum...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Rip_P.carotovorum...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Rup_P.carotovorum...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Slaad_P.carotovorum...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Crus_P.carotovorum...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Gander_P.carotovorum...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Guf_P.carotovorum...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Hoejben_P.carotovorum...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Magnum_P.carotovorum...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Vims_P.carotovorum...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Echoes_P.carotovorum...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Galvinrad_P.carotovorum...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Uther_P.carotovorum...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Mimer_P.carotovorum...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Abuela_P.carotovorum...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Koroua_P.carotovorum...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Sabo_P.carotovorum...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Taid_P.carotovorum...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Ymer_P.carotovorum...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Amona_P.carotovorum...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_carotovorum/+Poppous_P.carotovorum...\n",
      "\n",
      "Processing host range data for bacteria species: Pectobacterium atrosepticum\n",
      "Seq IDs for selected species: ['J10_21_reoriented', 'J11_21_reoriented', 'J126_23_reoriented', 'J12_21_reoriented', 'J16_21_reoriented', 'J22_21_reoriented', 'J28_21_reoriented', 'J33_21_reoriented', 'J38_21_reoriented', 'J39_21_reoriented', 'J40_21_reoriented', 'J56_21_reoriented', 'J6_21_reoriented', 'J7_21_reoriented', 'J8_21_reoriented', 'J116_23_reoriented', 'J118_23_reoriented', 'J121_23_reoriented', 'J125_23_reoriented', 'J70_21_reoriented', 'J78_22_reoriented', 'J81_22_reoriented', 'J82_22_reoriented', 'J84_22_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 1, 'Taid': 1, 'Poppous': 1, 'Koroua': 1, 'Abuela': 1, 'Amona': 1, 'Sabo': 0, 'Mimer': 1, 'Crus': 1, 'Gander': 1, 'Guf': 1, 'Hoejben': 1, 'Magnum': 1, 'Vims': 1, 'Echoes': 0, 'Galvinrad': 0, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 1, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Pantea_P.atrosepticum...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Rap_P.atrosepticum...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Zann_P.atrosepticum...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Rip_P.atrosepticum...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Rup_P.atrosepticum...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Slaad_P.atrosepticum...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Crus_P.atrosepticum...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Gander_P.atrosepticum...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Guf_P.atrosepticum...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Hoejben_P.atrosepticum...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Magnum_P.atrosepticum...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Vims_P.atrosepticum...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Echoes_P.atrosepticum...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Galvinrad_P.atrosepticum...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Uther_P.atrosepticum...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Mimer_P.atrosepticum...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Abuela_P.atrosepticum...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Koroua_P.atrosepticum...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Sabo_P.atrosepticum...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Taid_P.atrosepticum...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Ymer_P.atrosepticum...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Amona_P.atrosepticum...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_atrosepticum/+Poppous_P.atrosepticum...\n",
      "\n",
      "Processing host range data for bacteria species: Pectobacterium punjabense\n",
      "Seq IDs for selected species: ['J2_21_reoriented', 'J3_21_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 0, 'Taid': 0, 'Poppous': 0, 'Koroua': 0, 'Abuela': 0, 'Amona': 0, 'Sabo': 0, 'Mimer': 0, 'Crus': 0, 'Gander': 0, 'Guf': 0, 'Hoejben': 0, 'Magnum': 0, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 0, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 0, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Pantea_P.punjabense...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Rap_P.punjabense...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Zann_P.punjabense...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Rip_P.punjabense...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Rup_P.punjabense...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Slaad_P.punjabense...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Crus_P.punjabense...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Gander_P.punjabense...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Guf_P.punjabense...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Hoejben_P.punjabense...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Magnum_P.punjabense...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Vims_P.punjabense...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Echoes_P.punjabense...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Galvinrad_P.punjabense...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Uther_P.punjabense...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Mimer_P.punjabense...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Abuela_P.punjabense...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Koroua_P.punjabense...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Sabo_P.punjabense...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Taid_P.punjabense...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Ymer_P.punjabense...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Amona_P.punjabense...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_punjabense/+Poppous_P.punjabense...\n",
      "\n",
      "Processing host range data for bacteria species: Morganella morganii\n",
      "Seq IDs for selected species: ['J51_21_reoriented', 'J57_21_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 0, 'Taid': 0, 'Poppous': 0, 'Koroua': 0, 'Abuela': 0, 'Amona': 0, 'Sabo': 0, 'Mimer': 0, 'Crus': 0, 'Gander': 0, 'Guf': 0, 'Hoejben': 0, 'Magnum': 0, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 0, 'Uther': 0, 'Rip': 1, 'Rup': 1, 'Slaad': 1, 'Pantea': 0, 'Rap': 0, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Pantea_M.morganii...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Rap_M.morganii...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Zann_M.morganii...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Rip_M.morganii...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Rup_M.morganii...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Slaad_M.morganii...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Crus_M.morganii...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Gander_M.morganii...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Guf_M.morganii...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Hoejben_M.morganii...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Magnum_M.morganii...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Vims_M.morganii...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Echoes_M.morganii...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Galvinrad_M.morganii...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Uther_M.morganii...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Mimer_M.morganii...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Abuela_M.morganii...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Koroua_M.morganii...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Sabo_M.morganii...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Taid_M.morganii...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Ymer_M.morganii...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Amona_M.morganii...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Morganella_morganii/+Poppous_M.morganii...\n",
      "\n",
      "Processing host range data for bacteria species: Chishuiella\n",
      "Seq IDs for selected species: ['J105_22_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 0, 'Taid': 0, 'Poppous': 0, 'Koroua': 0, 'Abuela': 0, 'Amona': 0, 'Sabo': 0, 'Mimer': 0, 'Crus': 0, 'Gander': 0, 'Guf': 0, 'Hoejben': 0, 'Magnum': 0, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 0, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 0, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Pantea_Chishuiella...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Rap_Chishuiella...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Zann_Chishuiella...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Rip_Chishuiella...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Rup_Chishuiella...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Slaad_Chishuiella...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Crus_Chishuiella...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Gander_Chishuiella...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Guf_Chishuiella...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Hoejben_Chishuiella...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Magnum_Chishuiella...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Vims_Chishuiella...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Echoes_Chishuiella...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Galvinrad_Chishuiella...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Uther_Chishuiella...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Mimer_Chishuiella...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Abuela_Chishuiella...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Koroua_Chishuiella...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Sabo_Chishuiella...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Taid_Chishuiella...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Ymer_Chishuiella...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Amona_Chishuiella...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chishuiella/+Poppous_Chishuiella...\n",
      "\n",
      "Processing host range data for bacteria species: Pectobacterium brasiliense\n",
      "Seq IDs for selected species: ['J119_23_reoriented', 'J127_23_reoriented', 'J13_21_reoriented', 'J20_21_reoriented', 'J21_21_reoriented', 'J24_21_reoriented', 'J25_21_reoriented', 'J26_21_reoriented', 'J31_21_reoriented', 'J32_21_reoriented', 'J34_21_reoriented', 'J36_21_reoriented', 'J37_21_reoriented', 'J41_21_reoriented', 'J44_21_reoriented', 'J45_21_reoriented', 'J47_21_reoriented', 'J52_21_reoriented', 'J59_21_reoriented', 'J115_23_reoriented', 'J117_23_reoriented', 'J128_23_reoriented', 'J134_23_reoriented', 'J3009LEM_22_KMC_reoriented', 'J3009_2_22_KMC_reoriented', 'J69_21_reoriented', 'J73_21_reoriented', 'J77_22_reoriented', 'J86_22_reoriented', 'J97_22_reoriented', 'J98_22_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 1, 'Taid': 1, 'Poppous': 1, 'Koroua': 1, 'Abuela': 1, 'Amona': 1, 'Sabo': 1, 'Mimer': 1, 'Crus': 1, 'Gander': 1, 'Guf': 1, 'Hoejben': 1, 'Magnum': 1, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 0, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 1, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Pantea_P.brasiliense...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Rap_P.brasiliense...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Zann_P.brasiliense...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Rip_P.brasiliense...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Rup_P.brasiliense...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Slaad_P.brasiliense...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Crus_P.brasiliense...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Gander_P.brasiliense...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Guf_P.brasiliense...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Hoejben_P.brasiliense...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Magnum_P.brasiliense...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Vims_P.brasiliense...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Echoes_P.brasiliense...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Galvinrad_P.brasiliense...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Uther_P.brasiliense...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Mimer_P.brasiliense...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Abuela_P.brasiliense...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Koroua_P.brasiliense...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Sabo_P.brasiliense...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Taid_P.brasiliense...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Ymer_P.brasiliense...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Amona_P.brasiliense...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_brasiliense/+Poppous_P.brasiliense...\n",
      "\n",
      "Processing host range data for bacteria species: Serratia plymuthica\n",
      "Seq IDs for selected species: ['J4_21_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 0, 'Taid': 0, 'Poppous': 0, 'Koroua': 0, 'Abuela': 0, 'Amona': 0, 'Sabo': 0, 'Mimer': 0, 'Crus': 0, 'Gander': 0, 'Guf': 0, 'Hoejben': 0, 'Magnum': 0, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 1, 'Uther': 1, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 0, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Pantea_S.plymuthica...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Rap_S.plymuthica...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Zann_S.plymuthica...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Rip_S.plymuthica...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Rup_S.plymuthica...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Slaad_S.plymuthica...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Crus_S.plymuthica...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Gander_S.plymuthica...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Guf_S.plymuthica...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Hoejben_S.plymuthica...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Magnum_S.plymuthica...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Vims_S.plymuthica...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Echoes_S.plymuthica...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Galvinrad_S.plymuthica...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Uther_S.plymuthica...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Mimer_S.plymuthica...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Abuela_S.plymuthica...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Koroua_S.plymuthica...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Sabo_S.plymuthica...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Taid_S.plymuthica...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Ymer_S.plymuthica...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Amona_S.plymuthica...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Serratia_plymuthica/+Poppous_S.plymuthica...\n",
      "\n",
      "Processing host range data for bacteria species: Pectobacterium parmentieri\n",
      "Seq IDs for selected species: ['J27_21_reoriented', 'J29_21_reoriented', 'J30_21_reoriented', 'J42_21_reoriented', 'J49_21_reoriented', 'FO3A_23_KMC_reoriented', 'J103_22_reoriented', 'J104_22_reoriented', 'J106_22_reoriented', 'J108_22_reoriented', 'J62_22_reoriented', 'J65_21_reoriented', 'J68_21_reoriented', 'J83_22_reoriented', 'J93_22_reoriented', 'J95_22_reoriented', 'J96_22_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 1, 'Taid': 1, 'Poppous': 1, 'Koroua': 1, 'Abuela': 1, 'Amona': 1, 'Sabo': 0, 'Mimer': 1, 'Crus': 1, 'Gander': 1, 'Guf': 1, 'Hoejben': 1, 'Magnum': 1, 'Vims': 1, 'Echoes': 0, 'Galvinrad': 1, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 1, 'Rap': 0, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Pantea_P.parmentieri...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Rap_P.parmentieri...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Zann_P.parmentieri...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Rip_P.parmentieri...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Rup_P.parmentieri...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Slaad_P.parmentieri...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Crus_P.parmentieri...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Gander_P.parmentieri...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Guf_P.parmentieri...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Hoejben_P.parmentieri...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Magnum_P.parmentieri...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Vims_P.parmentieri...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Echoes_P.parmentieri...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Galvinrad_P.parmentieri...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Uther_P.parmentieri...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Mimer_P.parmentieri...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Abuela_P.parmentieri...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Koroua_P.parmentieri...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Sabo_P.parmentieri...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Taid_P.parmentieri...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Ymer_P.parmentieri...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Amona_P.parmentieri...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_parmentieri/+Poppous_P.parmentieri...\n",
      "\n",
      "Processing host range data for bacteria species: Pectobacterium polaris\n",
      "Seq IDs for selected species: ['J17_21_reoriented', 'J18_21_reoriented', 'J19_21_reoriented', 'J48_21_reoriented', 'J3009_1_22_KMC_reoriented', 'J60_21_reoriented', 'J61_21_reoriented', 'J66_21_reoriented', 'J75_21_reoriented', 'J76_21_reoriented', 'J79_22_reoriented', 'J94_22_reoriented', 'J99_22_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 1, 'Taid': 1, 'Poppous': 1, 'Koroua': 1, 'Abuela': 1, 'Amona': 1, 'Sabo': 0, 'Mimer': 1, 'Crus': 1, 'Gander': 1, 'Guf': 1, 'Hoejben': 1, 'Magnum': 1, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 1, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 1, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Pantea_P.polaris...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Rap_P.polaris...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Zann_P.polaris...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Rip_P.polaris...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Rup_P.polaris...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Slaad_P.polaris...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Crus_P.polaris...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Gander_P.polaris...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Guf_P.polaris...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Hoejben_P.polaris...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Magnum_P.polaris...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Vims_P.polaris...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Echoes_P.polaris...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Galvinrad_P.polaris...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Uther_P.polaris...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Mimer_P.polaris...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Abuela_P.polaris...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Koroua_P.polaris...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Sabo_P.polaris...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Taid_P.polaris...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Ymer_P.polaris...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Amona_P.polaris...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pectobacterium_polaris/+Poppous_P.polaris...\n",
      "\n",
      "Processing host range data for bacteria species: Pseudomonas chlororaphis\n",
      "Seq IDs for selected species: ['J109_23_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 0, 'Taid': 0, 'Poppous': 0, 'Koroua': 0, 'Abuela': 0, 'Amona': 0, 'Sabo': 0, 'Mimer': 0, 'Crus': 0, 'Gander': 0, 'Guf': 0, 'Hoejben': 0, 'Magnum': 0, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 0, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 0, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Pantea_P.chlororaphis...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Rap_P.chlororaphis...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Zann_P.chlororaphis...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Rip_P.chlororaphis...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Rup_P.chlororaphis...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Slaad_P.chlororaphis...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Crus_P.chlororaphis...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Gander_P.chlororaphis...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Guf_P.chlororaphis...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Hoejben_P.chlororaphis...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Magnum_P.chlororaphis...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Vims_P.chlororaphis...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Echoes_P.chlororaphis...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Galvinrad_P.chlororaphis...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Uther_P.chlororaphis...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Mimer_P.chlororaphis...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Abuela_P.chlororaphis...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Koroua_P.chlororaphis...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Sabo_P.chlororaphis...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Taid_P.chlororaphis...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Ymer_P.chlororaphis...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Amona_P.chlororaphis...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Pseudomonas_chlororaphis/+Poppous_P.chlororaphis...\n",
      "\n",
      "Processing host range data for bacteria species: Chryseobacterium\n",
      "Seq IDs for selected species: ['J46_21_reoriented', 'J50_21_reoriented', 'J2264_1_22_KMC_reoriented', 'J2264_3_22_KMC_reoriented', 'J63_22_reoriented', 'J64_22_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 0, 'Taid': 0, 'Poppous': 0, 'Koroua': 0, 'Abuela': 0, 'Amona': 0, 'Sabo': 0, 'Mimer': 0, 'Crus': 0, 'Gander': 0, 'Guf': 0, 'Hoejben': 0, 'Magnum': 0, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 0, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 0, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Pantea_Chryseobacterium...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Rap_Chryseobacterium...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Zann_Chryseobacterium...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Rip_Chryseobacterium...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Rup_Chryseobacterium...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Slaad_Chryseobacterium...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Crus_Chryseobacterium...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Gander_Chryseobacterium...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Guf_Chryseobacterium...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Hoejben_Chryseobacterium...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Magnum_Chryseobacterium...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Vims_Chryseobacterium...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Echoes_Chryseobacterium...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Galvinrad_Chryseobacterium...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Uther_Chryseobacterium...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Mimer_Chryseobacterium...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Abuela_Chryseobacterium...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Koroua_Chryseobacterium...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Sabo_Chryseobacterium...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Taid_Chryseobacterium...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Ymer_Chryseobacterium...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Amona_Chryseobacterium...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Chryseobacterium/+Poppous_Chryseobacterium...\n",
      "\n",
      "Processing host range data for bacteria species: Vagococcus\n",
      "Seq IDs for selected species: ['J5_21_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 0, 'Taid': 0, 'Poppous': 0, 'Koroua': 0, 'Abuela': 0, 'Amona': 0, 'Sabo': 0, 'Mimer': 0, 'Crus': 0, 'Gander': 0, 'Guf': 0, 'Hoejben': 0, 'Magnum': 0, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 0, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 0, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Pantea_Vagococcus...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Rap_Vagococcus...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Zann_Vagococcus...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Rip_Vagococcus...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Rup_Vagococcus...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Slaad_Vagococcus...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Crus_Vagococcus...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Gander_Vagococcus...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Guf_Vagococcus...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Hoejben_Vagococcus...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Magnum_Vagococcus...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Vims_Vagococcus...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Echoes_Vagococcus...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Galvinrad_Vagococcus...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Uther_Vagococcus...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Mimer_Vagococcus...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Abuela_Vagococcus...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Koroua_Vagococcus...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Sabo_Vagococcus...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Taid_Vagococcus...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Ymer_Vagococcus...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Amona_Vagococcus...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Vagococcus/+Poppous_Vagococcus...\n",
      "\n",
      "Processing host range data for bacteria species: Acinetobacter calcoaceticus\n",
      "Seq IDs for selected species: ['J14_21_reoriented', 'J53_21_reoriented']\n",
      "Combined host range data for selected species: {'Ymer': 0, 'Taid': 0, 'Poppous': 0, 'Koroua': 0, 'Abuela': 0, 'Amona': 0, 'Sabo': 0, 'Mimer': 0, 'Crus': 0, 'Gander': 0, 'Guf': 0, 'Hoejben': 0, 'Magnum': 0, 'Vims': 0, 'Echoes': 0, 'Galvinrad': 0, 'Uther': 0, 'Rip': 0, 'Rup': 0, 'Slaad': 0, 'Pantea': 0, 'Rap': 0, 'Zann': 0}\n",
      "\n",
      "Loading kmers from phages\n",
      "Phage: Pantea, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Pantea_A.calcoaceticus...\n",
      "Phage: Rap, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Rap_A.calcoaceticus...\n",
      "Phage: Zann, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Zann_A.calcoaceticus...\n",
      "Phage: Rip, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Rip_A.calcoaceticus...\n",
      "Phage: Rup, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Rup_A.calcoaceticus...\n",
      "Phage: Slaad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Slaad_A.calcoaceticus...\n",
      "Phage: Crus, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Crus_A.calcoaceticus...\n",
      "Phage: Gander, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Gander_A.calcoaceticus...\n",
      "Phage: Guf, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Guf_A.calcoaceticus...\n",
      "Phage: Hoejben, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Hoejben_A.calcoaceticus...\n",
      "Phage: Magnum, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Magnum_A.calcoaceticus...\n",
      "Phage: Vims, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Vims_A.calcoaceticus...\n",
      "Phage: Echoes, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Echoes_A.calcoaceticus...\n",
      "Phage: Galvinrad, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Galvinrad_A.calcoaceticus...\n",
      "Phage: Uther, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Uther_A.calcoaceticus...\n",
      "Phage: Mimer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Mimer_A.calcoaceticus...\n",
      "Phage: Abuela, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Abuela_A.calcoaceticus...\n",
      "Phage: Koroua, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Koroua_A.calcoaceticus...\n",
      "Phage: Sabo, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Sabo_A.calcoaceticus...\n",
      "Phage: Taid, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Taid_A.calcoaceticus...\n",
      "Phage: Ymer, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Ymer_A.calcoaceticus...\n",
      "Phage: Amona, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Amona_A.calcoaceticus...\n",
      "Phage: Poppous, Number of kmers: 886272\n",
      "Writing ../data_prod/phage_kmers_12_txt/Acinetobacter_calcoaceticus/+Poppous_A.calcoaceticus...\n"
     ]
    }
   ],
   "source": [
    "import shutil\n",
    "from manipulations import hostrange_bact\n",
    "\n",
    "parent_out_dir = data_prod_path + f\"phage_kmers_{K}_txt/\"\n",
    "if not os.path.exists(parent_out_dir):\n",
    "    os.makedirs(parent_out_dir)\n",
    "else:\n",
    "    shutil.rmtree(parent_out_dir)\n",
    "    os.makedirs(parent_out_dir)\n",
    "\n",
    "for selected_bact_species in set(bact_lookup.values()): \n",
    "    ### PREPPING HOST RANGE DATA ###\n",
    "    # Select a specific bacteria species for host range analysis\n",
    "    print(f\"\\nProcessing host range data for bacteria species: {selected_bact_species}\")\n",
    "\n",
    "    #obtain all the keys where the value is equal to selected_bact_species\n",
    "    selected_seqIDs = [key for key, value in bact_lookup.items() if value == selected_bact_species]\n",
    "    print(\"Seq IDs for selected species:\", selected_seqIDs)\n",
    "\n",
    "    # Acceptive approach: since all seqIDs for the same species should have similar host ranges, we combine their host range data.\n",
    "    # if non-zero is found for any seqID, we set it to 1 in the final host range data.\n",
    "\n",
    "    combined_host_range = hostrange_bact(host_range_data, selected_seqIDs, approach=\"acceptive\")\n",
    "    print(\"Combined host range data for selected species:\", combined_host_range)\n",
    "\n",
    "    ### LOADING PHAGE KMERS ###\n",
    "    phage_data = {}\n",
    "    print(f\"\\nLoading kmers from phages\")\n",
    "    for p_index in all_phage_kmers.index:\n",
    "        phage_name = p_index.split(\"_\")[-1]\n",
    "        curr_phage_kmers = all_phage_kmers.T[p_index]\n",
    "\n",
    "        if phage_name in combined_host_range:\n",
    "            phage_data[phage_name] = curr_phage_kmers.index\n",
    "        else:\n",
    "            print(f\"Warning: No matching phage name for {phage_name} in host range data. Skipping...\")\n",
    "\n",
    "    ### OUTPUTTING KMER TXT FILES ###\n",
    "    out_dir = parent_out_dir + f\"{str(selected_bact_species).replace(' ', '_')}/\"\n",
    "    if not os.path.exists(out_dir):\n",
    "        os.makedirs(out_dir)\n",
    "\n",
    "    for phage_name in phage_data.keys():\n",
    "        print(f\"Phage: {phage_name}, Number of kmers: {len(phage_data[phage_name])}\")\n",
    "        print(f\"Writing {out_dir}+{phage_name}_{short_species_name(selected_bact_species)}...\")\n",
    "        with open(os.path.join(out_dir, f\"{phage_name}.txt\"), 'w') as f:\n",
    "            for kmer in phage_data[phage_name]:\n",
    "                f.write(f\"{kmer}\\t{combined_host_range[phage_name]}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5670980b",
   "metadata": {},
   "source": [
    "### Make combined txt files per bacteria\n",
    "each file will contain which kmers has been in a phage that could infect the bacteria."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "21677a63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Combined kmers for Chryseobacterium into ../data_prod/phage_kmers_12_combined/Chryseobacterium_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Lelliottia into ../data_prod/phage_kmers_12_combined/Lelliottia_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Acinetobacter_calcoaceticus into ../data_prod/phage_kmers_12_combined/Acinetobacter_calcoaceticus_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Chishuiella into ../data_prod/phage_kmers_12_combined/Chishuiella_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Vagococcus into ../data_prod/phage_kmers_12_combined/Vagococcus_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Pectobacterium_polaris into ../data_prod/phage_kmers_12_combined/Pectobacterium_polaris_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Pseudomonas_chlororaphis into ../data_prod/phage_kmers_12_combined/Pseudomonas_chlororaphis_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Morganella_morganii into ../data_prod/phage_kmers_12_combined/Morganella_morganii_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Pectobacterium_parmentieri into ../data_prod/phage_kmers_12_combined/Pectobacterium_parmentieri_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Pectobacterium_brasiliense into ../data_prod/phage_kmers_12_combined/Pectobacterium_brasiliense_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Pectobacterium_atrosepticum into ../data_prod/phage_kmers_12_combined/Pectobacterium_atrosepticum_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Pseudomonas_marginalis into ../data_prod/phage_kmers_12_combined/Pseudomonas_marginalis_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Pectobacterium_carotovorum into ../data_prod/phage_kmers_12_combined/Pectobacterium_carotovorum_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Serratia_plymuthica into ../data_prod/phage_kmers_12_combined/Serratia_plymuthica_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Pectobacterium_punjabense into ../data_prod/phage_kmers_12_combined/Pectobacterium_punjabense_combined.txt\n",
      "Shuffling combined file for random distribution...\n",
      "Combined kmers for Serratia_liquefaciens into ../data_prod/phage_kmers_12_combined/Serratia_liquefaciens_combined.txt\n",
      "Shuffling combined file for random distribution...\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "### Concatenate all phage kmer txt files into a single file for easier NN training ###\n",
    "all_phage_kmers_txt_path = data_prod_path + f\"phage_kmers_{K}_combined/\"\n",
    "if not os.path.exists(all_phage_kmers_txt_path):\n",
    "    os.makedirs(all_phage_kmers_txt_path)\n",
    "\n",
    "for filename in os.listdir(parent_out_dir):\n",
    "    if os.path.isdir(os.path.join(parent_out_dir, filename)):\n",
    "        bact_folder = os.path.join(parent_out_dir, filename)\n",
    "        combined_output_file = os.path.join(all_phage_kmers_txt_path, f\"{filename}_combined.txt\")\n",
    "        with open(combined_output_file, 'w') as outfile:\n",
    "            for phage_file in os.listdir(bact_folder):\n",
    "                if phage_file.endswith('.txt'):\n",
    "                    phage_filepath = os.path.join(bact_folder, phage_file)\n",
    "                    with open(phage_filepath, 'r') as infile:\n",
    "                        shutil.copyfileobj(infile, outfile)\n",
    "        print(f\"Combined kmers for {filename} into {combined_output_file}\")\n",
    "        \n",
    "        print(\"Shuffling combined file for random distribution...\")\n",
    "        with open(combined_output_file, 'r') as f:\n",
    "            lines = f.readlines()\n",
    "        random.shuffle(lines)\n",
    "        with open(combined_output_file, 'w') as f:\n",
    "            f.writelines(lines)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c488da3",
   "metadata": {},
   "source": [
    "## Running NN"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9280c448",
   "metadata": {},
   "source": [
    "Inits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "31607ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Packages -----------------------------\n",
    "import random, shutil, os, sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import pickle\n",
    "import sys\n",
    "import pdb\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "### Paths --------------------------------\n",
    "raw_data_path = \"../raw_data/\"\n",
    "data_prod_path = \"../data_prod/\"\n",
    "NN_files_path = data_prod_path + \"NN_files/\"\n",
    "if not os.path.exists(NN_files_path):\n",
    "    os.makedirs(NN_files_path)\n",
    "else:\n",
    "    shutil.rmtree(NN_files_path)\n",
    "    os.makedirs(NN_files_path)\n",
    "\n",
    "### Custom variables ---------------------\n",
    "K = 12 #kmer size; equal to 6 aa.\n",
    "selected_bact_species = \"Pectobacterium brasiliense\" \n",
    "full_path = data_prod_path+f\"phage_kmers_{K}_combined/{selected_bact_species.replace(' ', '_')}_combined.txt\"\n",
    "if not os.path.exists(full_path):\n",
    "    raise FileNotFoundError(\"Combined kmer txt file for the selected bacteria species not found. Please run the data preparation steps first.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8606db7",
   "metadata": {},
   "source": [
    "### Splitting data\n",
    "each folder in phage_minhash_K_txt is a bacteria name, specifying whether the sketches of each underlying phage txt file, can infect it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f64af29d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def separate_test(path, train_val_ratio=0.9, test_ratio=0.1, TS = False) -> tuple:\n",
    "    \"\"\"\n",
    "    Separate the input file into training/validation and test sets based on specified ratios.\n",
    "    Write the separated data to respective train_val and test files.\n",
    "    Returns the paths to the train_val and test files.\n",
    "    \"\"\"\n",
    "    if abs(train_val_ratio + test_ratio - 1.0) > 1e-6:\n",
    "        raise ValueError(\"Train, validation, and test ratios must sum to 1.\")\n",
    "\n",
    "    if type(path) == list:\n",
    "        lines = []\n",
    "        for input_file in path:\n",
    "            if TS: print(f\"Separating data from {input_file} into train/val and test sets...\")\n",
    "            try: \n",
    "                with open(input_file, 'r') as f:\n",
    "                    lines.extend(f.readlines())\n",
    "            except FileNotFoundError as e:\n",
    "                print(\"File not found: \", e)\n",
    "    else:\n",
    "        if TS: print(f\"Separating data from {path} into train/val and test sets...\")\n",
    "        with open(path, 'r') as f:\n",
    "            lines = f.readlines()\n",
    "\n",
    "    random.shuffle(lines)\n",
    "\n",
    "    total_lines = len(lines)\n",
    "    train_val_end = int(total_lines * train_val_ratio)\n",
    "\n",
    "    train_data = lines[:train_val_end]\n",
    "    test_data = lines[train_val_end:]\n",
    "    \n",
    "    if TS: \n",
    "        print(f\"Total lines: {total_lines}, Train/Val lines: {len(train_data)}, Test lines: {len(test_data)}\")\n",
    "        print(f\"Ratios - Train/Val: {len(train_data)/total_lines:.2f}, Test: {len(test_data)/total_lines:.2f}\", end=\"\\n\\n\")\n",
    "\n",
    "    # Write the separated data to respective files to save memory\n",
    "    train_val_path = NN_files_path + f\"{selected_bact_species.replace(' ', '_')}_train_val.txt\"\n",
    "    with open(NN_files_path + f\"{selected_bact_species.replace(' ', '_')}_train_val.txt\", 'w') as f:\n",
    "        f.writelines(train_data)\n",
    "    \n",
    "    test_path = NN_files_path + f\"{selected_bact_species.replace(' ', '_')}_test.txt\"\n",
    "    with open(NN_files_path + f\"{selected_bact_species.replace(' ', '_')}_test.txt\", 'w') as f:\n",
    "        f.writelines(test_data)\n",
    "    \n",
    "    return train_val_path, test_path\n",
    "\n",
    "def lines_to_df(lines):\n",
    "    rows = [l.strip().split('\\t') for l in lines if l.strip()]\n",
    "    df = pd.DataFrame(rows, columns=['kmer', 'label'])\n",
    "    # set proper dtypes\n",
    "    df['kmer'] = df['kmer'].astype(str)\n",
    "    df['label'] = pd.to_numeric(df['label'], errors='coerce').astype(int)\n",
    "    return df\n",
    "\n",
    "def k_fold_split(train_val_file, k=5, seed=42, TS = False):\n",
    "    \"\"\"\n",
    "    Perform k-fold cross-validation split on the input file.\n",
    "    Returns a list of (train_data, val_data) tuples for each fold.\n",
    "    Does not return test data.\n",
    "    \"\"\"    \n",
    "    if TS: print(\"Initiating k_fold_split() -----\")\n",
    "    # Read all the shuffled train val data\n",
    "    with open(train_val_file, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "\n",
    "    # Compute fold sizes\n",
    "    total = len(lines)\n",
    "    fold_size = total // k\n",
    "    folds = [lines[i*fold_size:(i+1)*fold_size] for i in range(k)]\n",
    "    if TS: print(f\"total lines {total}, fold_size {fold_size}\")\n",
    "\n",
    "    # Handle remainder lines (if not evenly divisible)\n",
    "    remainder = lines[k*fold_size:]\n",
    "    for i, line in enumerate(remainder):\n",
    "        folds[i % k].append(line)\n",
    "\n",
    "    # Generate (train, val) pairs\n",
    "    all_folds = []\n",
    "    for i in range(k):\n",
    "        if TS: print(f\"\\nStarting split {i+1}/{k}...\")\n",
    "        val_lines = folds[i]\n",
    "        train_lines = [line for j, f in enumerate(folds) if j != i for line in f]\n",
    "\n",
    "        train_df = lines_to_df(train_lines)\n",
    "        if TS: \n",
    "            print(f\"train_df constructed with shape: {train_df.shape} & data types\\n{train_df.dtypes}\")\n",
    "            print(\"snippet of train_df:\")\n",
    "            display(train_df.head())\n",
    "            \n",
    "        val_df = lines_to_df(val_lines)\n",
    "        if TS: \n",
    "            print(f\"val_df constructed with shape: {val_df.shape} & data types\\n{val_df.dtypes}\")\n",
    "            print(\"snippet of val_df:\")\n",
    "            display(val_df.head())\n",
    "\n",
    "        all_folds.append((train_df, val_df))\n",
    "    \n",
    "    return all_folds\n",
    "\n",
    "def train_network(net, x_train, y_train, learning_rate):\n",
    "    \"\"\"\n",
    "    Trains the network for a single epoch, running the forward and backward pass, and compute and return the loss.\n",
    "    \"\"\"\n",
    "    # Forward pass\n",
    "    z1, a1, z2, a2  = net.forward(x_train)\n",
    "    # backward pass\n",
    "    backward(net, x_train, y_train, z1, a1, z2, a2, learning_rate)\n",
    "    loss = np.mean((a2 - y_train) ** 2)\n",
    "    return loss\n",
    "        \n",
    "def eval_network(net, x_valid, y_valid):\n",
    "    \"\"\"\n",
    "    Evaluates the network ; Note that we do not update weights (no backward pass)\n",
    "    \"\"\"\n",
    "    z1, a1, z2, a2 = net.forward(x_valid)\n",
    "    loss = np.mean((a2-y_valid)**2)\n",
    "    return loss\n",
    "\n",
    "# Model saving and loading functions\n",
    "def save_ffnn_model(filepath, model):\n",
    "    if not filepath.endswith('.pkl'):\n",
    "        filepath = filepath+'.pkl'\n",
    "    with open(filepath, 'wb') as f:\n",
    "        dict_to_save = {'input_size': model.W1.shape[0], 'hidden_size':model.W1.shape[1], 'output_size':model.W2.shape[1],\n",
    "                        'W1': model.W1, 'b1':model.b1, 'W2':model.W2, 'b2':model.b2}\n",
    "        pickle.dump(dict_to_save, f)\n",
    "        print(f'Saved FFNN model at {filepath}')\n",
    "\n",
    "\n",
    "def load_ffnn_model(filepath, model=None):\n",
    "\n",
    "    with open(filepath, 'rb') as f:\n",
    "        loaded_dict = pickle.load(f)\n",
    "    if model is None:\n",
    "            model = SimpleFFNN(loaded_dict['input_size'], loaded_dict['hidden_size'], loaded_dict['output_size'])\n",
    "    assert (model.W1.shape[0]==loaded_dict['input_size'] and model.W1.shape[1]==loaded_dict['hidden_size'] and model.W2.shape[1]==loaded_dict['output_size']), \\\n",
    "        f\"Model and loaded weights size mismatch!. Provided model has weight of dimensions {model.W1.shape, model.W2.shape} ; Loaded weights have shape {loaded_dict['W1'].shape, loaded_dict['W2'].shape}\"\n",
    "\n",
    "    model.W1 = loaded_dict['W1']\n",
    "    model.b1 = loaded_dict['b1']\n",
    "    model.W2 = loaded_dict['W2']\n",
    "    model.b2 = loaded_dict['b2']\n",
    "    print(f\"Model loaded successfully from {filepath}\\nwith weights [ W1, W2 ] dimensions : {model.W1.shape, model.W2.shape}\")\n",
    "    return model\n",
    "\n",
    "def plot_losses(train_losses, valid_losses, n_epochs, title=None):\n",
    "    # Plotting the losses \n",
    "    fig,ax = plt.subplots(1,1, figsize=(9,5))\n",
    "    ax.plot(range(n_epochs), train_losses, label='Train loss', c='b')\n",
    "    ax.plot(range(n_epochs), valid_losses, label='Valid loss', c='m')\n",
    "    ax.legend()\n",
    "    if title is not None:\n",
    "        fig.suptitle(title)\n",
    "    fig.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b57dbb1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Separating data from ../data_prod/phage_kmers_12_combined/Pectobacterium_brasiliense_combined.txt into train/val and test sets...\n",
      "Total lines: 20384256, Train/Val lines: 18345830, Test lines: 2038426\n",
      "Ratios - Train/Val: 0.90, Test: 0.10\n",
      "\n",
      "Training/Validation data saved to: ../data_prod/NN_files/Pectobacterium_brasiliense_train_val.txt\n",
      "Test data saved to: ../data_prod/NN_files/Pectobacterium_brasiliense_test.txt\n"
     ]
    }
   ],
   "source": [
    "train_val_path, test_path = separate_test(full_path, train_val_ratio=0.9, test_ratio=0.1, TS=True)\n",
    "print(f\"Training/Validation data saved to: {train_val_path}\")\n",
    "print(f\"Test data saved to: {test_path}\")\n",
    "\n",
    "#Retun list of paths were files start with Pecto in data_prod_path+f\"phage_minhash_{K}_combined/\n",
    "# combined_path = data_prod_path+f\"phage_minhash_{K}_combined/\"\n",
    "# pecto_files = [combined_path+path for path in os.listdir(combined_path) if path.startswith(\"Pecto\")]\n",
    "# print(pecto_files)\n",
    "# train_val_path, test_path = separate_test(pecto_files, train_val_ratio=0.9, test_ratio=0.1, TS=True)\n",
    "# print(f\"Training/Validation data saved to: {train_val_path}\")\n",
    "# print(f\"Test data saved to: {test_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d94a1918",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Purely to define split, no writing\n",
    "#k_fold_data = k_fold_split(train_val_path, k=5, seed=42)\n",
    "#for fold_idx, (train_data, val_data) in enumerate(k_fold_data):\n",
    "    #print(f\"Fold {fold_idx+1}: Train data length: {len(train_data)}, Val data length: {len(val_data)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4b8888f",
   "metadata": {},
   "source": [
    "### Defining NN class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bc332f50",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Weights initialization function.\n",
    "def xavier_initialization_normal(input_dim, output_dim):\n",
    "    shape = (input_dim, output_dim)\n",
    "    stddev = np.sqrt(2 / (input_dim + output_dim))\n",
    "    return np.random.normal(0, stddev, size=shape) * 0.1\n",
    "\n",
    "def random_initialization_normal(input_dim, output_dim):\n",
    "    return np.random.randn(input_dim, output_dim) * 0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "266061c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "class SimpleFFNN:\n",
    "    def __init__(self, input_size, hidden_size, output_size, initialization_function=xavier_initialization_normal):\n",
    "        # Initialize weights and biases with small random values\n",
    "        # initialization_function(input_dim, output_dim) -> np.array of shape (input_dim, output_dim)\n",
    "        self.W1 = initialization_function(input_size, hidden_size)\n",
    "        self.b1 = np.zeros(hidden_size)\n",
    "        self.W2 = initialization_function(hidden_size, output_size)\n",
    "        self.b2 = np.zeros(output_size)\n",
    "        print(f'Input -> Hidden Layer Weight Matrix Shape: {self.W1.shape}',\n",
    "              f'First Layer Bias Weights Vector Shape: {self.b1.shape}',\n",
    "              f'Hidden -> Output layer Weight Matrix Shape: {self.W2.shape}',\n",
    "              f'Second Layer Bias Weights Vector Shape: {self.b2.shape}', sep=\"\\n\")\n",
    "        \n",
    "    def relu(self, x):\n",
    "        return np.maximum(0, x)\n",
    "\n",
    "    def sigmoid(self, x): \n",
    "        \"\"\"\n",
    "        The normal version of sigmoid 1 / (1 + np.exp(-x)) is NOT numerically stable\n",
    "        Here we split the case into two for positive and negative inputs\n",
    "        because np.exp(-x) for something negative will quickly overflow if x is a large negative number\n",
    "        \"\"\"\n",
    "        # This is equivalent to : \n",
    "        # if x>=0, then compute (1/(1+np.exp(-x)))\n",
    "        # if x<0: compute (np.exp(x)/(1+np.exp(x))))\n",
    "        return np.where(x >= 0, 1 / (1 + np.exp(-x)), \n",
    "                        np.exp(x) / (1 + np.exp(x)))\n",
    "\n",
    "    def forward(self, x):\n",
    "        \"\"\"\n",
    "        x:\n",
    "        zi denotes the output of a hidden layer i\n",
    "        ai denotes the output of an activation function (non-linearity) at layer i\n",
    "        (activations are relu, sigmoid, tanh, etc.)\n",
    "        Use self.function to call a method. for example: self.relu(XX)\n",
    "        \"\"\"\n",
    "\n",
    "        # First layer : Use a relu here for the activation \n",
    "        z1 = np.dot(x, self.W1) + self.b1 #XX\n",
    "        a1 = self.relu(z1) #XX\n",
    "        \n",
    "        # Output layer : Use a sigmoid here for the activation\n",
    "        z2 = np.dot(a1, self.W2) + self.b2 #XX\n",
    "        a2 = self.sigmoid(z2) #XX\n",
    "        \n",
    "        # Return all the intermediate outputs as well because we need them for backpropagation (see slides)\n",
    "        return z1, a1, z2, a2\n",
    "\n",
    "def relu_derivative(a):\n",
    "    return np.where(a > 0, 1, 0)\n",
    "\n",
    "def sigmoid_derivative(a):\n",
    "    \"\"\"\n",
    "    For this derivative, it is not necessary to find a numerically stable version.\n",
    "    Just take the base formula and derive it.    \n",
    "    \"\"\"\n",
    "    return np.array(a*(1 - a))\n",
    "\n",
    "\n",
    "def backward(net, x, y, z1, a1, z2, a2, learning_rate=0.01):\n",
    "    \"\"\"\n",
    "    Function to backpropagate the gradients from the output to update the weights.\n",
    "    Apply the chain rule and slowly work out the chain derivatives from the output back to the input\n",
    "    Reminder that np.dot(array_1, array_2) and array.T exists to transpose an array for matrix multiplication\n",
    "    \"\"\"\n",
    "    # This assumes that we are computing a MSE as the loss function.\n",
    "    # Look at your slides to compute the gradient backpropagation for a mean-squared error using the chain rule.\n",
    "\n",
    "    # Output layer error ; We used a sigmoid in this layer\n",
    "    dE_dO = a2 - y\n",
    "    dO_do = sigmoid_derivative(a2)\n",
    "    dE_do = dE_dO * dO_do\n",
    "\n",
    "    #print(\"dE_do\", dE_do.shape)\n",
    "    ### (REMEMBER for np.dot(A,B) columns of A MUST equal rows in B) ###\n",
    "    \n",
    "    # Backpropagate to hidden layer \n",
    "    #print(\"a1\", a1.shape)\n",
    "    dE_dW2 = np.dot(dE_do.T, a1)\n",
    "    dE_db2 = np.sum(dE_do, axis=0, keepdims=True)\n",
    "    dE_db2 = dE_db2.squeeze() # Squeeze is needed here to make the dimensions fit\n",
    "    #print(\"dE_dW2\", dE_dW2.shape)\n",
    "    #print(\"dE_db2\", dE_dW2.shape)\n",
    "\n",
    "    # Hidden layer error ; We used a ReLU in this layer!\n",
    "    # (O − t)⋅ g'(o)⋅wj\n",
    "    dE_dH = np.dot(dE_do, net.W2.T)\n",
    "    #print(\"dE_dH\", dE_dH.shape)\n",
    "    dE_dh = dE_dH * relu_derivative(a1)\n",
    "    #print(\"dE_dh\", dE_dh.shape)\n",
    "\n",
    "    # Backpropagate to input layer\n",
    "    dE_dW1 = np.dot(dE_dh.T, x)\n",
    "    #print(\"dE_dh\", dE_dh.shape)\n",
    "    dE_db1 = np.sum(dE_dh, axis=0, keepdims=True) \n",
    "    dE_db1 = dE_db1.squeeze() # Squeeze is needed here to make the dimensions fit\n",
    "    #print(\"dE_db1\", dE_db1.shape)\n",
    "\n",
    "    # Update weights and biases using gradient descent\n",
    "    net.W1 -= learning_rate * dE_dW1.T\n",
    "    #print(\"W1 shapes:\", net.W1.shape, dE_dW1.T.shape)\n",
    "    net.b1 -= learning_rate * dE_db1.T\n",
    "    #print(\"b1 shapes:\", net.b1.shape, dE_db1.T.shape)\n",
    "    net.W2 -= learning_rate * dE_dW2.T\n",
    "    #print(\"W2 shapes:\", net.W2.shape, dE_dW2.T.shape)\n",
    "    net.b2 -= learning_rate * dE_db2.T\n",
    "    #print(\"b2 shapes:\", net.b2.shape, dE_db2.T.shape)\n",
    "\n",
    "def OHE(X_in, max_pep_len=9):\n",
    "    \"\"\"\n",
    "    One-hot encoding; Each of the four DNA bases (A, C, G, T) is represented by a unique binary vector where only one element is '1' and the rest are '0'.\n",
    "    Returns a tensor of encoded nucleotides of shape (1, max_pep_len, n_features) for a single batch\n",
    "    \"\"\"\n",
    "    mapping = {\n",
    "        'A': [1, 0, 0, 0],\n",
    "        'C': [0, 1, 0, 0],\n",
    "        'G': [0, 0, 1, 0],\n",
    "        'T': [0, 0, 0, 1]\n",
    "    }\n",
    "    \n",
    "    batch_size = len(X_in)\n",
    "    n_features = len(mapping)\n",
    "    \n",
    "    X_out = np.zeros((batch_size, max_pep_len, n_features), dtype=np.int8)\n",
    "    \n",
    "    for nucleotides_index, row in X_in.iterrows():\n",
    "        for aa_index in range(len(row)):\n",
    "            X_out[nucleotides_index, aa_index] = mapping[ row.nucleotides[aa_index] ].values\n",
    "            \n",
    "    return X_out, np.expand_dims(X_in.target.values,1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24f7e71a",
   "metadata": {},
   "source": [
    "### Train & Validate model\n",
    "an encoder function is needed to parse my train and val data, in order to perform 2 crucial steps:\n",
    "1. Fixing Sequence Length (Padding and Truncation)\n",
    "2. Creating the 3D Tensor Structure; $$(\\text{Batch Size}, \\text{Sequence Length}, \\text{Feature Dimensions})$$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7426cfed",
   "metadata": {},
   "source": [
    "#### Encode data for model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1b7d78ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initiating k_fold_split() -----\n",
      "total lines 18345830, fold_size 3669166\n",
      "\n",
      "Starting split 1/5...\n",
      "train_df constructed with shape: (14676664, 2) & data types\n",
      "kmer     object\n",
      "label     int64\n",
      "dtype: object\n",
      "snippet of train_df:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCTGGAAGATTT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGACTGTTTTCT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGAGGTTAAACG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTCACCGCCATC</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACGCGTATGTCA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           kmer  label\n",
       "0  TCTGGAAGATTT      1\n",
       "1  TGACTGTTTTCT      1\n",
       "2  AGAGGTTAAACG      1\n",
       "3  TTCACCGCCATC      0\n",
       "4  ACGCGTATGTCA      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "val_df constructed with shape: (3669166, 2) & data types\n",
      "kmer     object\n",
      "label     int64\n",
      "dtype: object\n",
      "snippet of val_df:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCGACAACCGCA</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGTACGACCTTG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACCCTGACTATC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACGGCGGTTATT</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCAGTGTTTTCT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           kmer  label\n",
       "0  GCGACAACCGCA      0\n",
       "1  AGTACGACCTTG      0\n",
       "2  ACCCTGACTATC      1\n",
       "3  ACGGCGGTTATT      0\n",
       "4  TCAGTGTTTTCT      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Starting split 2/5...\n",
      "train_df constructed with shape: (14676664, 2) & data types\n",
      "kmer     object\n",
      "label     int64\n",
      "dtype: object\n",
      "snippet of train_df:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCGACAACCGCA</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGTACGACCTTG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACCCTGACTATC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACGGCGGTTATT</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCAGTGTTTTCT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           kmer  label\n",
       "0  GCGACAACCGCA      0\n",
       "1  AGTACGACCTTG      0\n",
       "2  ACCCTGACTATC      1\n",
       "3  ACGGCGGTTATT      0\n",
       "4  TCAGTGTTTTCT      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "val_df constructed with shape: (3669166, 2) & data types\n",
      "kmer     object\n",
      "label     int64\n",
      "dtype: object\n",
      "snippet of val_df:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCTGGAAGATTT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGACTGTTTTCT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGAGGTTAAACG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTCACCGCCATC</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACGCGTATGTCA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           kmer  label\n",
       "0  TCTGGAAGATTT      1\n",
       "1  TGACTGTTTTCT      1\n",
       "2  AGAGGTTAAACG      1\n",
       "3  TTCACCGCCATC      0\n",
       "4  ACGCGTATGTCA      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Starting split 3/5...\n",
      "train_df constructed with shape: (14676664, 2) & data types\n",
      "kmer     object\n",
      "label     int64\n",
      "dtype: object\n",
      "snippet of train_df:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCGACAACCGCA</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGTACGACCTTG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACCCTGACTATC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACGGCGGTTATT</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCAGTGTTTTCT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           kmer  label\n",
       "0  GCGACAACCGCA      0\n",
       "1  AGTACGACCTTG      0\n",
       "2  ACCCTGACTATC      1\n",
       "3  ACGGCGGTTATT      0\n",
       "4  TCAGTGTTTTCT      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "val_df constructed with shape: (3669166, 2) & data types\n",
      "kmer     object\n",
      "label     int64\n",
      "dtype: object\n",
      "snippet of val_df:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GGCTGTTCGTCG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GGTCCGGTTAAC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGTACCTACAAC</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TACGGTTACCTC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CGGCGTTACTTA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           kmer  label\n",
       "0  GGCTGTTCGTCG      1\n",
       "1  GGTCCGGTTAAC      1\n",
       "2  AGTACCTACAAC      0\n",
       "3  TACGGTTACCTC      1\n",
       "4  CGGCGTTACTTA      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Starting split 4/5...\n",
      "train_df constructed with shape: (14676664, 2) & data types\n",
      "kmer     object\n",
      "label     int64\n",
      "dtype: object\n",
      "snippet of train_df:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCGACAACCGCA</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGTACGACCTTG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACCCTGACTATC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACGGCGGTTATT</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCAGTGTTTTCT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           kmer  label\n",
       "0  GCGACAACCGCA      0\n",
       "1  AGTACGACCTTG      0\n",
       "2  ACCCTGACTATC      1\n",
       "3  ACGGCGGTTATT      0\n",
       "4  TCAGTGTTTTCT      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "val_df constructed with shape: (3669166, 2) & data types\n",
      "kmer     object\n",
      "label     int64\n",
      "dtype: object\n",
      "snippet of val_df:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGTCAGAGCAT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAACGAGCAAAA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATGATAATGTTG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCAAACGTGAGC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAAGCCAGAATA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           kmer  label\n",
       "0  TCGTCAGAGCAT      1\n",
       "1  GAACGAGCAAAA      1\n",
       "2  ATGATAATGTTG      1\n",
       "3  TCAAACGTGAGC      1\n",
       "4  CAAGCCAGAATA      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Starting split 5/5...\n",
      "train_df constructed with shape: (14676664, 2) & data types\n",
      "kmer     object\n",
      "label     int64\n",
      "dtype: object\n",
      "snippet of train_df:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCGACAACCGCA</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGTACGACCTTG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACCCTGACTATC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACGGCGGTTATT</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCAGTGTTTTCT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           kmer  label\n",
       "0  GCGACAACCGCA      0\n",
       "1  AGTACGACCTTG      0\n",
       "2  ACCCTGACTATC      1\n",
       "3  ACGGCGGTTATT      0\n",
       "4  TCAGTGTTTTCT      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "val_df constructed with shape: (3669166, 2) & data types\n",
      "kmer     object\n",
      "label     int64\n",
      "dtype: object\n",
      "snippet of val_df:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCAGCAGATACA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAATATGGCTAT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGGTGCGGTCG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CGTTCATCTGAC</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATCTAAGGCGTG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           kmer  label\n",
       "0  CCAGCAGATACA      1\n",
       "1  CAATATGGCTAT      1\n",
       "2  TCGGTGCGGTCG      0\n",
       "3  CGTTCATCTGAC      0\n",
       "4  ATCTAAGGCGTG      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "k_fold_data = k_fold_split(train_val_path, k=5, seed=42, TS=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "76cd6aef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fold 1: Train data length: 14676664, Val data length: 3669166\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>label</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5742702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8933962</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         count\n",
       "label         \n",
       "0      5742702\n",
       "1      8933962"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "AttributeError",
     "evalue": "'Series' object has no attribute 'nucleotides'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/sd/ygh_pqf56ygdrztjkzy96_s00000gn/T/ipykernel_6576/1625277486.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mfold_idx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mtrain_data\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval_data\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk_fold_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Fold {fold_idx+1}: Train data length: {len(train_data)}, Val data length: {len(val_data)}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mdisplay\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtrain_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"label\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0magg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'kmer'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'count'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m     \u001b[0;31m#Encoding data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m     \u001b[0mx_train_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_train_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mOHE\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtrain_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0mx_valid_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_valid_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mOHE\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx_train_\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/sd/ygh_pqf56ygdrztjkzy96_s00000gn/T/ipykernel_6576/2568724674.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(X_in, max_pep_len)\u001b[0m\n\u001b[1;32m    124\u001b[0m     \u001b[0mX_out\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mzeros\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbatch_size\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_pep_len\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_features\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mint8\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    125\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    126\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mnucleotides_index\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mX_in\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miterrows\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    127\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0maa_index\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 128\u001b[0;31m             \u001b[0mX_out\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mnucleotides_index\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maa_index\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmapping\u001b[0m\u001b[0;34m[\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnucleotides\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maa_index\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    129\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    130\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mX_out\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexpand_dims\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_in\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtarget\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6295\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6296\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6297\u001b[0m         ):\n\u001b[1;32m   6298\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6299\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'Series' object has no attribute 'nucleotides'"
     ]
    }
   ],
   "source": [
    "#Purely to define split, no writing\n",
    "for fold_idx, (train_data, val_data) in enumerate(k_fold_data):\n",
    "    print(f\"Fold {fold_idx+1}: Train data length: {len(train_data)}, Val data length: {len(val_data)}\")\n",
    "    display(train_data.groupby(\"label\").agg(count=('kmer','count')))\n",
    "\n",
    "    #Encoding data\n",
    "    x_train_, y_train_ = OHE(train_data)\n",
    "    x_valid_, y_valid_ = OHE(val_data)\n",
    "    print(x_train_.shape)\n",
    "    print(x_train_)\n",
    "\n",
    "    #Initializing model\n",
    "    input_size = x_train_.shape[1] # also known as \"n_features\"\n",
    "    # Model and training hyperparameters\n",
    "    learning_rate = 0.0001\n",
    "    hidden_units = 50\n",
    "    n_epochs = 500\n",
    "    output_size = 1\n",
    "    # Creating a model instance \n",
    "    # You can use either `xavier_initialization_normal` or `random_initialization_normal`\n",
    "    # for the initialization_function argument of the class\n",
    "    network = SimpleFFNN(input_size, hidden_units, output_size)#, \n",
    "                        #initialization_function=xavier_initialization_normal)\n",
    "\n",
    "    # Training loops\n",
    "    train_losses = []\n",
    "    valid_losses = []\n",
    "\n",
    "    # Run n_epochs of training\n",
    "    for epoch in range(n_epochs):\n",
    "        train_loss = train_network(network, x_train_, y_train_, learning_rate)\n",
    "        valid_loss = eval_network(network, x_valid_, y_valid_)\n",
    "        train_losses.append(train_loss)\n",
    "        valid_losses.append(valid_loss)\n",
    "        # For the first, every 5% of the epochs and last epoch, we print the loss \n",
    "        # to check that the model is properly training. (loss going down)\n",
    "        if (n_epochs >= 10 and epoch % math.ceil(0.05 * n_epochs) == 0) or epoch == 0 or epoch == n_epochs:\n",
    "            print(f\"Epoch {epoch}: Train Loss: {train_loss:.4f}\\tValid Loss: {valid_loss:.4f}\")\n",
    "\n",
    "    # save model + plot losses (put your own savename to be used for the model and predictions)\n",
    "    model_savepath = NN_files_path+'some_ffnn_model.pkl' # /path/to/your/stuff/filename.pkl\n",
    "    save_ffnn_model(model_savepath, model=network)\n",
    "\n",
    "    # plotting the losses \n",
    "    plot_losses(train_losses, valid_losses, n_epochs, title=f\"Train & Val loss for {selected_bact_species} with {K}mers\")\n",
    "\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81fbcf01",
   "metadata": {},
   "source": [
    "### Testing model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2603cde8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
